{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from varseek.utils import vcf_to_dataframe\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "import varseek\n",
    "import numpy as np\n",
    "tqdm.pandas()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig '19' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse_filter] FILTER 'FAIL' is not defined in the header\n",
      "01:47:23 - INFO - Using the seq_id_column:var_column 'chromosome:mutation_from_vcf' columns as the variant header column.\n",
      "01:47:23 - INFO - Removing 0 duplications > k\n",
      "/Users/joeyrich/Desktop/local/varseek/varseek/varseek_build.py:1415: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  mutations[\"vcrs_sequence_kmer_length\"] = mutations[\"vcrs_sequence\"].apply(lambda x: len(x) if pd.notna(x) else 0)\n",
      "01:47:23 - INFO - Removed 1 variant kmers with length less than 59...\n",
      "01:47:23 - INFO - Removed 0 variant kmers containing more than 0 'N's...\n",
      "01:47:23 - WARNING - \n",
      "        12 variants correctly recorded (75.00%)\n",
      "        4 variants removed (25.00%)\n",
      "          0 variants missing seq_id or var_column (0.000%)\n",
      "          0 entries removed due to having a duplicate entry (0.000%)\n",
      "          0 variants with seq_ID not found in sequences (0.000%)\n",
      "          0 intronic variants found (0.000%)\n",
      "          0 posttranslational region variants found (0.000%)\n",
      "          1 unknown variants found (6.250%)\n",
      "          0 variants with uncertain mutation found (0.000%)\n",
      "          0 variants with ambiguous position found (0.000%)\n",
      "          0 variants with incorrect wildtype base found (0.000%)\n",
      "          0 variants with indices outside of the sequence length found (0.000%)\n",
      "          0 duplications longer than k found (0.000%)\n",
      "          2 variants with overlapping kmers found (12.500%)\n",
      "          1 variants with fragment length < min_seq_len removed (6.250%)\n",
      "          0 variants with Ns found (0.000%)\n",
      "        \n",
      "01:47:23 - INFO - Merging rows of identical VCRSs\n",
      "01:47:23 - INFO - \n",
      "        Number of variants total: 12\n",
      "        Number of variants merged: 0\n",
      "        Number of unique variants: 12\n",
      "        Number of VCRSs: 12\n",
      "        \n",
      "01:47:23 - INFO - Merged headers were combined and separated using a semicolon (;). Occurences of identical VCRSs may be reduced by increasing w.\n",
      "01:47:23 - INFO - FASTA file containing VCRSs created at vcf_testing_out_dir/vcrs.fa.\n",
      "01:47:23 - INFO - t2g file containing VCRSs created at vcf_testing_out_dir/vcrs_t2g.txt.\n",
      "01:47:23 - INFO - Total runtime for vk build: 0m, 26.80s\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "vcf_testing_out_dir/vcrs_t2g.txt already exists\n"
     ]
    }
   ],
   "source": [
    "import varseek as vk\n",
    "vk.build(\n",
    "    variants=\"vcf_testing.vcf\",\n",
    "    sequences=\"sequences_testing.fa\",\n",
    "    out=\"vcf_testing_out_dir\",\n",
    "    seq_id_column=\"chromosome\",\n",
    "    var_column=\"mutation_from_vcf\",\n",
    "    overwrite=True\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !vk build -v vcf_testing.vcf -s sequences_testing.fa -o vcf_testing_out_dir --seq_id_column chromosome --var_column mutation_from_vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig '19' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse_filter] FILTER 'FAIL' is not defined in the header\n",
      "[W::vcf_parse] Contig '' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::bcf_hrec_check] Invalid contig name: \"\"\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Error extracting ID",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[2], line 6\u001b[0m\n\u001b[1;32m      3\u001b[0m var_id_column \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mvariant_id\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m      4\u001b[0m var_column \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmutation_genome\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m----> 6\u001b[0m mutations \u001b[38;5;241m=\u001b[39m \u001b[43mvcf_to_dataframe\u001b[49m\u001b[43m(\u001b[49m\u001b[43msample_vcf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43madditional_columns\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mexplode_alt\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mfilter_empty_alt\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m      7\u001b[0m \u001b[38;5;66;03m# sample_vcf_df['CHROM'] = sample_vcf_df['CHROM'].astype(str)\u001b[39;00m\n\u001b[1;32m      8\u001b[0m mutations\u001b[38;5;241m.\u001b[39mrename(columns\u001b[38;5;241m=\u001b[39m{\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCHROM\u001b[39m\u001b[38;5;124m\"\u001b[39m: seq_id_column}, inplace\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m~/Desktop/local/varseek/varseek/utils/seq_utils.py:945\u001b[0m, in \u001b[0;36mvcf_to_dataframe\u001b[0;34m(vcf_file, additional_columns, explode_alt, filter_empty_alt)\u001b[0m\n\u001b[1;32m    938\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m record \u001b[38;5;129;01min\u001b[39;00m vcf\u001b[38;5;241m.\u001b[39mfetch():\n\u001b[1;32m    939\u001b[0m     \u001b[38;5;66;03m# For each record, extract the desired fields\u001b[39;00m\n\u001b[1;32m    940\u001b[0m     alts \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m,\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mjoin(record\u001b[38;5;241m.\u001b[39malts) \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(record\u001b[38;5;241m.\u001b[39malts, \u001b[38;5;28mtuple\u001b[39m) \u001b[38;5;28;01melse\u001b[39;00m record\u001b[38;5;241m.\u001b[39malts  \u001b[38;5;66;03m# alternate case includes None (when it is simply \".\")\u001b[39;00m\n\u001b[1;32m    942\u001b[0m     vcf_row \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    943\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCHROM\u001b[39m\u001b[38;5;124m\"\u001b[39m: record\u001b[38;5;241m.\u001b[39mchrom,\n\u001b[1;32m    944\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mPOS\u001b[39m\u001b[38;5;124m\"\u001b[39m: record\u001b[38;5;241m.\u001b[39mpos,\n\u001b[0;32m--> 945\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mID\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[43mrecord\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mid\u001b[49m,\n\u001b[1;32m    946\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mREF\u001b[39m\u001b[38;5;124m\"\u001b[39m: record\u001b[38;5;241m.\u001b[39mref,\n\u001b[1;32m    947\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mALT\u001b[39m\u001b[38;5;124m\"\u001b[39m: alts,  \u001b[38;5;66;03m# ALT can be multiple\u001b[39;00m\n\u001b[1;32m    948\u001b[0m     }\n\u001b[1;32m    950\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m additional_columns:\n\u001b[1;32m    951\u001b[0m         vcf_row[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mQUAL\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m record\u001b[38;5;241m.\u001b[39mqual\n",
      "File \u001b[0;32m~/miniconda3/envs/varseek/lib/python3.10/site-packages/pysam/libcbcf.pyx:3179\u001b[0m, in \u001b[0;36mpysam.libcbcf.VariantRecord.id.__get__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Error extracting ID"
     ]
    }
   ],
   "source": [
    "sample_vcf = \"vcf_testing.vcf\"  # \"/Users/joeyrich/Downloads/sample.vcf\"\n",
    "seq_id_column = \"chromosome\"\n",
    "var_id_column = \"variant_id\"\n",
    "var_column = \"mutation_genome\"\n",
    "\n",
    "mutations = vcf_to_dataframe(sample_vcf, additional_columns=True, explode_alt=True, filter_empty_alt=True)\n",
    "# sample_vcf_df['CHROM'] = sample_vcf_df['CHROM'].astype(str)\n",
    "mutations.rename(columns={\"CHROM\": seq_id_column}, inplace=True)\n",
    "if var_id_column:\n",
    "    mutations.rename(columns={\"ID\": var_id_column}, inplace=True)\n",
    "mutations[var_column] = mutations.progress_apply(generate_mutation_notation_from_vcf_columns, axis=1)  #!! untested"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "chromosome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "POS",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "variant_id",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "REF",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ALT",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_genome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "QUAL",
         "rawType": "float64",
         "type": "float"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "fe9df7fe-052b-4cfd-96db-0733a1f3b922",
       "rows": [
        [
         "0",
         "19",
         "11",
         "single_sub_mid;g.11A>C",
         "A",
         "C",
         "g.11A>C",
         "1.0"
        ],
        [
         "1",
         "19",
         "22",
         "exploded_sub_mid;g.22G>A,g.22G>C,g.22G>T",
         "G",
         "A",
         "g.22G>A",
         "1.0"
        ],
        [
         "2",
         "19",
         "22",
         "exploded_sub_mid;g.22G>A,g.22G>C,g.22G>T",
         "G",
         "C",
         "g.22G>C",
         "1.0"
        ],
        [
         "3",
         "19",
         "22",
         "exploded_sub_mid;g.22G>A,g.22G>C,g.22G>T",
         "G",
         "T",
         "g.22G>T",
         "1.0"
        ],
        [
         "4",
         "19",
         "1",
         "sub_begin;g.1T>A",
         "T",
         "A",
         "g.1T>A",
         "1.0"
        ],
        [
         "5",
         "19",
         "34",
         "exploded_ins_mid;g.34_35insA,g.34_35insGAC",
         "G",
         "GA",
         "g.34_35insA",
         "0.0"
        ],
        [
         "6",
         "19",
         "34",
         "exploded_ins_mid;g.34_35insA,g.34_35insGAC",
         "G",
         "GAC",
         "g.34_35insAC",
         "0.0"
        ],
        [
         "7",
         "19",
         "1",
         "ins_begin;None",
         "T",
         "AGT",
         "g.UNKNOWN",
         "0.0"
        ],
        [
         "8",
         "19",
         "40",
         "single_del_mid;g.40del",
         "AG",
         "A",
         "g.41del",
         "0.0"
        ],
        [
         "9",
         "19",
         "40",
         "multi_del_mid;g.40_43del",
         "AGCAT",
         "A",
         "g.41_44del",
         "0.0"
        ],
        [
         "10",
         "19",
         "1",
         "single_del_begin;g.1del",
         "TC",
         "C",
         "g.1del",
         "0.0"
        ],
        [
         "11",
         "19",
         "1",
         "multi_del_begin;g.1_4del",
         "TCATC",
         "C",
         "g.1_5del",
         "0.0"
        ],
        [
         "12",
         "19",
         "60",
         "delins;g.60_62delinsAA",
         "TCG",
         "AA",
         "g.60_62delinsAA",
         "0.0"
        ]
       ],
       "shape": {
        "columns": 7,
        "rows": 13
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>POS</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>mutation_genome</th>\n",
       "      <th>QUAL</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>single_sub_mid;g.11A&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>g.11A&gt;C</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>exploded_sub_mid;g.22G&gt;A,g.22G&gt;C,g.22G&gt;T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>g.22G&gt;A</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>exploded_sub_mid;g.22G&gt;A,g.22G&gt;C,g.22G&gt;T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>g.22G&gt;C</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>exploded_sub_mid;g.22G&gt;A,g.22G&gt;C,g.22G&gt;T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>g.22G&gt;T</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>sub_begin;g.1T&gt;A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>g.1T&gt;A</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>19</td>\n",
       "      <td>34</td>\n",
       "      <td>exploded_ins_mid;g.34_35insA,g.34_35insGAC</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>g.34_35insA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>19</td>\n",
       "      <td>34</td>\n",
       "      <td>exploded_ins_mid;g.34_35insA,g.34_35insGAC</td>\n",
       "      <td>G</td>\n",
       "      <td>GAC</td>\n",
       "      <td>g.34_35insAC</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>ins_begin;None</td>\n",
       "      <td>T</td>\n",
       "      <td>AGT</td>\n",
       "      <td>g.UNKNOWN</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>19</td>\n",
       "      <td>40</td>\n",
       "      <td>single_del_mid;g.40del</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>g.41del</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>19</td>\n",
       "      <td>40</td>\n",
       "      <td>multi_del_mid;g.40_43del</td>\n",
       "      <td>AGCAT</td>\n",
       "      <td>A</td>\n",
       "      <td>g.41_44del</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>single_del_begin;g.1del</td>\n",
       "      <td>TC</td>\n",
       "      <td>C</td>\n",
       "      <td>g.1del</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>multi_del_begin;g.1_4del</td>\n",
       "      <td>TCATC</td>\n",
       "      <td>C</td>\n",
       "      <td>g.1_5del</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>19</td>\n",
       "      <td>60</td>\n",
       "      <td>delins;g.60_62delinsAA</td>\n",
       "      <td>TCG</td>\n",
       "      <td>AA</td>\n",
       "      <td>g.60_62delinsAA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome  POS                                  variant_id    REF  ALT  \\\n",
       "0          19   11                      single_sub_mid;g.11A>C      A    C   \n",
       "1          19   22    exploded_sub_mid;g.22G>A,g.22G>C,g.22G>T      G    A   \n",
       "2          19   22    exploded_sub_mid;g.22G>A,g.22G>C,g.22G>T      G    C   \n",
       "3          19   22    exploded_sub_mid;g.22G>A,g.22G>C,g.22G>T      G    T   \n",
       "4          19    1                            sub_begin;g.1T>A      T    A   \n",
       "5          19   34  exploded_ins_mid;g.34_35insA,g.34_35insGAC      G   GA   \n",
       "6          19   34  exploded_ins_mid;g.34_35insA,g.34_35insGAC      G  GAC   \n",
       "7          19    1                              ins_begin;None      T  AGT   \n",
       "8          19   40                      single_del_mid;g.40del     AG    A   \n",
       "9          19   40                    multi_del_mid;g.40_43del  AGCAT    A   \n",
       "10         19    1                     single_del_begin;g.1del     TC    C   \n",
       "11         19    1                    multi_del_begin;g.1_4del  TCATC    C   \n",
       "12         19   60                      delins;g.60_62delinsAA    TCG   AA   \n",
       "\n",
       "    mutation_genome  QUAL  \n",
       "0           g.11A>C   1.0  \n",
       "1           g.22G>A   1.0  \n",
       "2           g.22G>C   1.0  \n",
       "3           g.22G>T   1.0  \n",
       "4            g.1T>A   1.0  \n",
       "5       g.34_35insA   0.0  \n",
       "6      g.34_35insAC   0.0  \n",
       "7         g.UNKNOWN   0.0  \n",
       "8           g.41del   0.0  \n",
       "9        g.41_44del   0.0  \n",
       "10           g.1del   0.0  \n",
       "11         g.1_5del   0.0  \n",
       "12  g.60_62delinsAA   0.0  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations[['chromosome', 'POS', 'variant_id', 'REF', 'ALT', 'mutation_genome', 'QUAL']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def add_variant_type_column_to_vcf_derived_df(sample_vcf_df):\n",
    "    # Compute lengths once\n",
    "    sample_vcf_df[\"REF_len\"] = sample_vcf_df[\"REF\"].str.len()\n",
    "    sample_vcf_df[\"ALT_len\"] = sample_vcf_df[\"ALT\"].str.len()\n",
    "\n",
    "    sample_vcf_df[\"ALT_RC\"] = sample_vcf_df[\"ALT\"].apply(varseek.utils.reverse_complement)\n",
    "\n",
    "    # Define conditions using precomputed values\n",
    "    #!!! check for duplications of length > 1 later\n",
    "    conditions = [\n",
    "        (sample_vcf_df[\"REF_len\"] == 1) & (sample_vcf_df[\"ALT_len\"] == 1),  # Substitution\n",
    "        (sample_vcf_df[\"REF_len\"] > 1) & (sample_vcf_df[\"ALT_len\"] == 1),   # Deletion\n",
    "        (sample_vcf_df[\"REF_len\"] == 1) & (sample_vcf_df[\"ALT_len\"] == 2) & (sample_vcf_df[\"ALT\"].str[1] == sample_vcf_df[\"REF\"].str[0]),  # Duplication of length 1 - must go before insertion because it is a special case of insertion\n",
    "        (sample_vcf_df[\"REF_len\"] == 1) & (sample_vcf_df[\"ALT_len\"] > 1),   # Insertion\n",
    "        (sample_vcf_df[\"REF_len\"] > 1) & (sample_vcf_df[\"ALT_len\"] > 1) & (sample_vcf_df[\"REF\"] == sample_vcf_df[\"ALT_RC\"]),  # Inversion - must go before delins because it is a special case of delins\n",
    "        (sample_vcf_df[\"REF_len\"] > 1) & (sample_vcf_df[\"ALT_len\"] > 1),    # Delins\n",
    "    ]\n",
    "\n",
    "    # Define corresponding values\n",
    "    choices = [\"substitution\", \"deletion\", \"duplication\", \"insertion\", \"inversion\", \"delins\"]\n",
    "\n",
    "    # Apply np.select\n",
    "    sample_vcf_df[\"variant_type\"] = np.select(conditions, choices, default=\"unknown\")\n",
    "    \n",
    "def add_variant_column_to_vcf_derived_df(sample_vcf_df, var_column=\"variant\"):\n",
    "    # Compute end position for delins\n",
    "    sample_vcf_df[\"start_POS_deletion\"] = (sample_vcf_df[\"POS\"] + 1).astype(str)\n",
    "    sample_vcf_df[\"start_POS_deletion_starting_at_1\"] = \"1\"\n",
    "    sample_vcf_df[\"end_POS_for_multibase_deletion_and_delins_and_inversion\"] = (sample_vcf_df[\"POS\"] + sample_vcf_df[\"REF_len\"] - 1).astype(str)\n",
    "    sample_vcf_df[\"end_POS_for_multibase_deletion_starting_at_1\"] = (sample_vcf_df[\"REF_len\"] - 1).astype(str)\n",
    "    sample_vcf_df[\"end_POS_for_insertion\"] = (sample_vcf_df[\"POS\"] + 1).astype(str)\n",
    "    sample_vcf_df[\"ALT_first_base_trimmed\"] = sample_vcf_df[\"ALT\"].str[1:]\n",
    "    # sample_vcf_df[\"ALT_last_base_trimmed\"] = sample_vcf_df[\"ALT\"].str[:-1]\n",
    "\n",
    "    # Define conditions\n",
    "    conditions = [\n",
    "        sample_vcf_df[\"variant_type\"] == \"substitution\",  # Substitution\n",
    "        (sample_vcf_df[\"variant_type\"] == \"deletion\") & (sample_vcf_df[\"REF_len\"] == 2) & (sample_vcf_df[\"POS\"] != 1),  # Single base deletion\n",
    "        (sample_vcf_df[\"variant_type\"] == \"deletion\") & (sample_vcf_df[\"REF_len\"] > 2) & (sample_vcf_df[\"POS\"] != 1),  # Multi base deletion\n",
    "        (sample_vcf_df[\"variant_type\"] == \"deletion\") & (sample_vcf_df[\"REF_len\"] == 2) & (sample_vcf_df[\"POS\"] == 1),  # Single base deletion starting at 1\n",
    "        (sample_vcf_df[\"variant_type\"] == \"deletion\") & (sample_vcf_df[\"REF_len\"] > 2) & (sample_vcf_df[\"POS\"] == 1),  # Multi base deletion starting at 1\n",
    "        (sample_vcf_df[\"variant_type\"] == \"insertion\") & (sample_vcf_df[\"POS\"] != 1),  # Insertion\n",
    "        (sample_vcf_df[\"variant_type\"] == \"insertion\") & (sample_vcf_df[\"POS\"] == 1),  # Insertion starting at 1\n",
    "        sample_vcf_df[\"variant_type\"] == \"delins\",  # Delins\n",
    "        sample_vcf_df[\"variant_type\"] == \"duplication\",  # Single base duplication\n",
    "        sample_vcf_df[\"variant_type\"] == \"inversion\"  # Inversion\n",
    "    ]\n",
    "\n",
    "    # Ensure POS is an integer in string format\n",
    "    sample_vcf_df[\"POS\"] = sample_vcf_df[\"POS\"].astype(str)\n",
    "\n",
    "    # Define corresponding variant formats\n",
    "    choices = [\n",
    "        \"g.\" + sample_vcf_df[\"POS\"] + sample_vcf_df[\"REF\"] + \">\" + sample_vcf_df[\"ALT\"],  # Substitution\n",
    "        \"g.\" + sample_vcf_df[\"start_POS_deletion\"] + \"del\",  # Single base deletion\n",
    "        \"g.\" + sample_vcf_df[\"start_POS_deletion\"] + \"_\" + sample_vcf_df[\"end_POS_for_multibase_deletion_and_delins_and_inversion\"] + \"del\",  # Multi base deletion\n",
    "        \"g.1del\",  # Single base deletion starting at 1\n",
    "        \"g.1_\" + sample_vcf_df[\"end_POS_for_multibase_deletion_starting_at_1\"] + \"del\",  # Multi base deletion starting at 1\n",
    "        \"g.\" + sample_vcf_df[\"POS\"] + \"_\" + sample_vcf_df[\"end_POS_for_insertion\"] + \"ins\" + sample_vcf_df[\"ALT_first_base_trimmed\"],  # Insertion\n",
    "        \"g.unknown\",  # Insertion starting at 1  # \"g.0_1ins\" + sample_vcf_df[\"ALT_last_base_trimmed\"],\n",
    "        \"g.\" + sample_vcf_df[\"POS\"] + \"_\" + sample_vcf_df[\"end_POS_for_multibase_deletion_and_delins_and_inversion\"] + \"delins\" + sample_vcf_df[\"ALT\"],  # Delins\n",
    "        \"g.\" + sample_vcf_df[\"POS\"] + \"dup\",  # Single base duplication\n",
    "        \"g.\" + sample_vcf_df[\"POS\"] + \"_\" + sample_vcf_df[\"end_POS_for_multibase_deletion_and_delins_and_inversion\"] + \"inv\"  # Inversion\n",
    "    ]\n",
    "\n",
    "    # Apply np.select\n",
    "    sample_vcf_df[var_column] = np.select(conditions, choices, default=\"g.unknown\")  # Default to None if no match\n",
    "    sample_vcf_df.drop(columns=[\"REF_len\", \"ALT_len\", \"ALT_RC\", \"start_POS_deletion\", \"start_POS_deletion_starting_at_1\", \"end_POS_for_multibase_deletion_and_delins_and_inversion\", \"end_POS_for_multibase_deletion_starting_at_1\", \"end_POS_for_insertion\", \"ALT_first_base_trimmed\", \"ALT_last_base_trimmed\"], inplace=True, errors=\"ignore\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig '19' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse_filter] FILTER 'FAIL' is not defined in the header\n"
     ]
    }
   ],
   "source": [
    "sample_vcf = \"vcf_testing.vcf\"  # \"/Users/joeyrich/Downloads/sample.vcf\"\n",
    "seq_id_column = \"chromosome\"\n",
    "var_id_column = \"variant_id\"\n",
    "var_column = \"mutation_genome\"\n",
    "\n",
    "sample_vcf_df = vcf_to_dataframe(sample_vcf, additional_columns=False, explode_alt=True, filter_empty_alt=True)\n",
    "# sample_vcf_df['CHROM'] = sample_vcf_df['CHROM'].astype(str)\n",
    "sample_vcf_df.rename(columns={\"CHROM\": seq_id_column}, inplace=True)\n",
    "if var_id_column:\n",
    "    sample_vcf_df.rename(columns={\"ID\": var_id_column}, inplace=True)\n",
    "\n",
    "add_variant_type_column_to_vcf_derived_df(sample_vcf_df)\n",
    "add_variant_column_to_vcf_derived_df(sample_vcf_df, var_column=var_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "index",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "chromosome",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "POS",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "variant_id",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "REF",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "ALT",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "variant_type",
         "rawType": "object",
         "type": "string"
        },
        {
         "name": "mutation_genome",
         "rawType": "object",
         "type": "string"
        }
       ],
       "conversionMethod": "pd.DataFrame",
       "ref": "0357a017-5439-4f61-acb9-9aedd8eefabb",
       "rows": [
        [
         "0",
         "19",
         "11",
         "single_sub_mid;g.11A>C",
         "A",
         "C",
         "substitution",
         "g.11A>C"
        ],
        [
         "1",
         "19",
         "22",
         "exploded_sub_mid;g.22G>A,g.22G>C,g.22G>T",
         "G",
         "A",
         "substitution",
         "g.22G>A"
        ],
        [
         "2",
         "19",
         "22",
         "exploded_sub_mid;g.22G>A,g.22G>C,g.22G>T",
         "G",
         "C",
         "substitution",
         "g.22G>C"
        ],
        [
         "3",
         "19",
         "22",
         "exploded_sub_mid;g.22G>A,g.22G>C,g.22G>T",
         "G",
         "T",
         "substitution",
         "g.22G>T"
        ],
        [
         "4",
         "19",
         "1",
         "sub_begin;g.1T>A",
         "T",
         "A",
         "substitution",
         "g.1T>A"
        ],
        [
         "5",
         "19",
         "34",
         "exploded_ins_mid;g.34_35insA,g.34_35insGAC",
         "G",
         "GA",
         "insertion",
         "g.34_35insA"
        ],
        [
         "6",
         "19",
         "34",
         "exploded_ins_mid;g.34_35insA,g.34_35insGAC",
         "G",
         "GAC",
         "insertion",
         "g.34_35insAC"
        ],
        [
         "7",
         "19",
         "1",
         "ins_begin;None",
         "T",
         "AGT",
         "insertion",
         "g.unknown"
        ],
        [
         "8",
         "19",
         "40",
         "single_del_mid;g.41del",
         "AG",
         "A",
         "deletion",
         "g.41del"
        ],
        [
         "9",
         "19",
         "40",
         "multi_del_mid;g.41_44del",
         "AGCAT",
         "A",
         "deletion",
         "g.41_44del"
        ],
        [
         "10",
         "19",
         "1",
         "single_del_begin;g.1del",
         "TC",
         "C",
         "deletion",
         "g.1del"
        ],
        [
         "11",
         "19",
         "1",
         "multi_del_begin;g.1_4del",
         "TCATC",
         "C",
         "deletion",
         "g.1_4del"
        ],
        [
         "12",
         "19",
         "60",
         "delins;g.60_62delinsAA",
         "TCG",
         "AA",
         "delins",
         "g.60_62delinsAA"
        ],
        [
         "13",
         "19",
         "40",
         "duplication;g.40dup",
         "A",
         "AA",
         "duplication",
         "g.40dup"
        ],
        [
         "14",
         "19",
         "40",
         "inversion;g.40_43inv",
         "TCGT",
         "ACGA",
         "inversion",
         "g.40_43inv"
        ],
        [
         "15",
         "19",
         "50",
         "duplication;g.47_50dup(initiallyg.50_51insGCCC)",
         "C",
         "CGCCC",
         "insertion",
         "g.50_51insGCCC"
        ]
       ],
       "shape": {
        "columns": 7,
        "rows": 16
       }
      },
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>POS</th>\n",
       "      <th>variant_id</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>variant_type</th>\n",
       "      <th>mutation_genome</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>11</td>\n",
       "      <td>single_sub_mid;g.11A&gt;C</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>substitution</td>\n",
       "      <td>g.11A&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>exploded_sub_mid;g.22G&gt;A,g.22G&gt;C,g.22G&gt;T</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>substitution</td>\n",
       "      <td>g.22G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>exploded_sub_mid;g.22G&gt;A,g.22G&gt;C,g.22G&gt;T</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>substitution</td>\n",
       "      <td>g.22G&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>22</td>\n",
       "      <td>exploded_sub_mid;g.22G&gt;A,g.22G&gt;C,g.22G&gt;T</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>substitution</td>\n",
       "      <td>g.22G&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>sub_begin;g.1T&gt;A</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>substitution</td>\n",
       "      <td>g.1T&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>19</td>\n",
       "      <td>34</td>\n",
       "      <td>exploded_ins_mid;g.34_35insA,g.34_35insGAC</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>insertion</td>\n",
       "      <td>g.34_35insA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>19</td>\n",
       "      <td>34</td>\n",
       "      <td>exploded_ins_mid;g.34_35insA,g.34_35insGAC</td>\n",
       "      <td>G</td>\n",
       "      <td>GAC</td>\n",
       "      <td>insertion</td>\n",
       "      <td>g.34_35insAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>ins_begin;None</td>\n",
       "      <td>T</td>\n",
       "      <td>AGT</td>\n",
       "      <td>insertion</td>\n",
       "      <td>g.unknown</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>19</td>\n",
       "      <td>40</td>\n",
       "      <td>single_del_mid;g.41del</td>\n",
       "      <td>AG</td>\n",
       "      <td>A</td>\n",
       "      <td>deletion</td>\n",
       "      <td>g.41del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>19</td>\n",
       "      <td>40</td>\n",
       "      <td>multi_del_mid;g.41_44del</td>\n",
       "      <td>AGCAT</td>\n",
       "      <td>A</td>\n",
       "      <td>deletion</td>\n",
       "      <td>g.41_44del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>single_del_begin;g.1del</td>\n",
       "      <td>TC</td>\n",
       "      <td>C</td>\n",
       "      <td>deletion</td>\n",
       "      <td>g.1del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>19</td>\n",
       "      <td>1</td>\n",
       "      <td>multi_del_begin;g.1_4del</td>\n",
       "      <td>TCATC</td>\n",
       "      <td>C</td>\n",
       "      <td>deletion</td>\n",
       "      <td>g.1_4del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>19</td>\n",
       "      <td>60</td>\n",
       "      <td>delins;g.60_62delinsAA</td>\n",
       "      <td>TCG</td>\n",
       "      <td>AA</td>\n",
       "      <td>delins</td>\n",
       "      <td>g.60_62delinsAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>19</td>\n",
       "      <td>40</td>\n",
       "      <td>duplication;g.40dup</td>\n",
       "      <td>A</td>\n",
       "      <td>AA</td>\n",
       "      <td>duplication</td>\n",
       "      <td>g.40dup</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>19</td>\n",
       "      <td>40</td>\n",
       "      <td>inversion;g.40_43inv</td>\n",
       "      <td>TCGT</td>\n",
       "      <td>ACGA</td>\n",
       "      <td>inversion</td>\n",
       "      <td>g.40_43inv</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>19</td>\n",
       "      <td>50</td>\n",
       "      <td>duplication;g.47_50dup(initiallyg.50_51insGCCC)</td>\n",
       "      <td>C</td>\n",
       "      <td>CGCCC</td>\n",
       "      <td>insertion</td>\n",
       "      <td>g.50_51insGCCC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome POS                                       variant_id    REF  \\\n",
       "0          19  11                           single_sub_mid;g.11A>C      A   \n",
       "1          19  22         exploded_sub_mid;g.22G>A,g.22G>C,g.22G>T      G   \n",
       "2          19  22         exploded_sub_mid;g.22G>A,g.22G>C,g.22G>T      G   \n",
       "3          19  22         exploded_sub_mid;g.22G>A,g.22G>C,g.22G>T      G   \n",
       "4          19   1                                 sub_begin;g.1T>A      T   \n",
       "5          19  34       exploded_ins_mid;g.34_35insA,g.34_35insGAC      G   \n",
       "6          19  34       exploded_ins_mid;g.34_35insA,g.34_35insGAC      G   \n",
       "7          19   1                                   ins_begin;None      T   \n",
       "8          19  40                           single_del_mid;g.41del     AG   \n",
       "9          19  40                         multi_del_mid;g.41_44del  AGCAT   \n",
       "10         19   1                          single_del_begin;g.1del     TC   \n",
       "11         19   1                         multi_del_begin;g.1_4del  TCATC   \n",
       "12         19  60                           delins;g.60_62delinsAA    TCG   \n",
       "13         19  40                              duplication;g.40dup      A   \n",
       "14         19  40                             inversion;g.40_43inv   TCGT   \n",
       "15         19  50  duplication;g.47_50dup(initiallyg.50_51insGCCC)      C   \n",
       "\n",
       "      ALT  variant_type  mutation_genome  \n",
       "0       C  substitution          g.11A>C  \n",
       "1       A  substitution          g.22G>A  \n",
       "2       C  substitution          g.22G>C  \n",
       "3       T  substitution          g.22G>T  \n",
       "4       A  substitution           g.1T>A  \n",
       "5      GA     insertion      g.34_35insA  \n",
       "6     GAC     insertion     g.34_35insAC  \n",
       "7     AGT     insertion        g.unknown  \n",
       "8       A      deletion          g.41del  \n",
       "9       A      deletion       g.41_44del  \n",
       "10      C      deletion           g.1del  \n",
       "11      C      deletion         g.1_4del  \n",
       "12     AA        delins  g.60_62delinsAA  \n",
       "13     AA   duplication          g.40dup  \n",
       "14   ACGA     inversion       g.40_43inv  \n",
       "15  CGCCC     insertion   g.50_51insGCCC  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutations = sample_vcf_df.copy()\n",
    "sample_vcf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_dict = {\"19\": \"TCATCGAACTAGCAGCTCGACGACGCACATCGTGGATCCAGCATCAGCCCCCTCTCGAGTCGCATCGCATCG\"}\n",
    "mutations[\"wt_sequence_full\"] = mutations[seq_id_column].map(seq_dict)\n",
    "mutations[\"ALT_len\"] = mutations[\"ALT\"].str.len()\n",
    "mutations[\"ALT_first_base_trimmed\"] = mutations[\"ALT\"].str[1:]\n",
    "\n",
    "# Step 1: Create a mask for rows where variant_type == \"insertion\" and the last base of ALT equals REF\n",
    "mask = (mutations[\"variant_type\"] == \"insertion\") & (mutations[\"ALT\"].str[-1] == mutations[\"REF\"])\n",
    "\n",
    "# Step 2: Compute the start position for comparison (only for masked rows)\n",
    "mutations.loc[mask, \"start_pos\"] = mutations.loc[mask, \"POS\"].astype(int) - mutations.loc[mask, \"ALT_len\"] + 2\n",
    "\n",
    "# Step 3: Extract sequence slice only for masked rows\n",
    "mutations.loc[mask, \"seq_slice\"] = mutations.loc[mask].apply(\n",
    "    lambda row: row[\"wt_sequence_full\"][int(row[\"start_pos\"]-1):int(row[\"POS\"])]  # row[\"start_pos\"]+1 because of 0-indexing in python\n",
    "    if pd.notna(row[\"wt_sequence_full\"]) else \"\", axis=1\n",
    ")\n",
    "\n",
    "# Step 4: Create a mask for matching sequences\n",
    "compare_mask = mask & (mutations[\"ALT_first_base_trimmed\"] == mutations[\"seq_slice\"])\n",
    "\n",
    "# Step 5: Update variant_type and variant only for matched rows\n",
    "mutations.loc[compare_mask, \"variant_type\"] = \"duplication\"\n",
    "mutations.loc[compare_mask, var_column] = \"g.\" + mutations.loc[compare_mask, \"start_pos\"].astype(int).astype(str) + \"_\" + mutations.loc[compare_mask, \"POS\"].astype(str) + \"dup\"\n",
    "\n",
    "mutations.drop(columns=[\"wt_sequence_full\", \"ALT_first_base_trimmed\", \"ALT_len\", \"start_pos\", \"seq_slice\"], inplace=True, errors=\"ignore\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "varseek",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
