{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     mutation   gene mutation_cdna\n",
      "0     p.R175H   TP53      c.524G>A\n",
      "1  c.5266dupC  BRCA1    c.5266dupC\n",
      "2     p.L858R   EGFR     c.2573T>G\n",
      "3     p.V600E   BRAF              \n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Example original DataFrame\n",
    "df_original = pd.DataFrame({\n",
    "    'mutation': ['p.R175H', 'c.5266dupC', 'p.L858R', 'p.V600E'],\n",
    "    'gene': ['TP53', 'BRCA1', 'EGFR', 'BRAF']\n",
    "})\n",
    "\n",
    "# Example df with mutation_cdna mapping\n",
    "df = pd.DataFrame({\n",
    "    'mutation': ['p.R175H', 'c.5266dupC', 'p.L858R'],\n",
    "    'mutation_cdna': ['c.524G>A', 'c.5266dupC', 'c.2573T>G']\n",
    "})\n",
    "\n",
    "# Merge on 'mutation' column, keeping all rows from df_original\n",
    "df_merged = df_original.merge(df[['mutation', 'mutation_cdna']], on='mutation', how='left')\n",
    "\n",
    "# Fill missing mutation_cdna values with an empty string if needed\n",
    "df_merged['mutation_cdna'] = df_merged['mutation_cdna'].fillna('')\n",
    "\n",
    "print(df_merged)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Time taken: 1.6109421253204346 seconds\n"
     ]
    }
   ],
   "source": [
    "import time\n",
    "start_time = time.time()\n",
    "variants = [\"1\"] * 1_000_000_000\n",
    "end_time = time.time()\n",
    "print(f\"Time taken: {end_time - start_time} seconds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    Gene    Mutation  Frequency\n",
      "0   TP53     p.R175H       0.30\n",
      "1  BRCA1  c.5266dupC       0.15\n",
      "2   EGFR     p.L858R       0.40\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Create a toy DataFrame\n",
    "variants = pd.DataFrame({\n",
    "    'Gene': ['TP53', 'BRCA1', 'EGFR'],\n",
    "    'Mutation': ['p.R175H', 'c.5266dupC', 'p.L858R'],\n",
    "    'Frequency': [0.3, 0.15, 0.4]\n",
    "})\n",
    "\n",
    "# Rename the variable\n",
    "mutations = variants\n",
    "del variants  # Optional\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations[\"test\"] = mutations[\"Frequency\"] * 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations[\"Frequency\"] = mutations[\"Frequency\"].astype(\"float32\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on function ref in module varseek.varseek_ref:\n",
      "\n",
      "ref(sequences, mutations, filters=('dlist_substring:equal=none', 'pseudoaligned_to_human_reference_despite_not_truly_aligning:is_not_true', 'dlist:equal=none', 'number_of_kmers_with_overlap_to_other_mcrs_items_in_mcrs_reference:less_than=999999', 'number_of_mcrs_items_with_overlapping_kmers_in_mcrs_reference:less_than=999999', 'longest_homopolymer_length:bottom_percent=99.99', 'triplet_complexity:top_percent=99.9'), mode=None, dlist=False, config=None, out='.', index_out=None, t2g_out=None, download=False, dry_run=False, list_downloadable_references=False, minimum_info_columns=True, overwrite=False, threads=2, verbose=True, **kwargs)\n",
      "        Create a reference index and t2g file for variant screening with varseek count. Wraps around varseek build, varseek info, varseek filter, and kb ref.\n",
      "    \n",
      "        # Required input argument:\n",
      "        - sequences     (str) Path to the fasta file containing the sequences to have the mutations added, e.g., 'seqs.fa'.\n",
      "                        Sequence identifiers following the '>' character must correspond to the identifiers\n",
      "                        in the seq_ID column of 'mutations'.\n",
      "    \n",
      "                        Example:\n",
      "                        >seq1 (or ENSG00000106443)\n",
      "                        ACTGCGATAGACT\n",
      "                        >seq2\n",
      "                        AGATCGCTAG\n",
      "    \n",
      "                        Alternatively: Input sequence(s) as a string or a list of strings,\n",
      "                        e.g. 'AGCTAGCT' or ['ACTGCTAGCT', 'AGCTAGCT'].\n",
      "    \n",
      "                        NOTE: Only the letters until the first space or dot will be used as sequence identifiers\n",
      "                        - Version numbers of Ensembl IDs will be ignored.\n",
      "                        NOTE: When 'sequences' input is a genome, also see 'gtf' argument below.\n",
      "    \n",
      "                        Alternatively, if 'mutations' is a string specifying a supported database, \n",
      "                        sequences can be a string indicating the source upon which to apply the mutations.\n",
      "                        See below for supported databases and sequences options.\n",
      "                        To see the supported combinations of mutations and sequences, either\n",
      "                        1) run `vk build --list_supported_databases` from the command line, or\n",
      "                        2) run varseek.build(list_supported_databases=True) in python\n",
      "    \n",
      "        - mutations     (str or DataFrame object) Path to csv or tsv file (str) (e.g., 'mutations.csv'), or DataFrame (DataFrame object),\n",
      "                        containing information about the mutations in the following format:\n",
      "    \n",
      "                        | mutation         | mut_ID | seq_ID |\n",
      "                        | c.2C>T           | mut1   | seq1   | -> Apply mutation 1 to sequence 1\n",
      "                        | c.9_13inv        | mut2   | seq2   | -> Apply mutation 2 to sequence 2\n",
      "                        | c.9_13inv        | mut2   | seq3   | -> Apply mutation 2 to sequence 3\n",
      "                        | c.9_13delinsAAT  | mut3   | seq3   | -> Apply mutation 3 to sequence 3\n",
      "                        | ...              | ...    | ...    |\n",
      "    \n",
      "                        'mutation' = Column containing the mutations to be performed written in standard mutation annotation (see below)\n",
      "                        'seq_ID' = Column containing the identifiers of the sequences to be mutated (must correspond to the string following\n",
      "                        the > character in the 'sequences' fasta file; do NOT include spaces or dots)\n",
      "                        'mut_ID' = Column containing an identifier for each mutation (optional).\n",
      "    \n",
      "                        Alternatively: Input mutation(s) as a string or list, e.g., 'c.2C>T' or ['c.2C>T', 'c.1A>C'].\n",
      "                        If a list is provided, the number of mutations must equal the number of input sequences.\n",
      "    \n",
      "                        For more information on the standard mutation annotation, see https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1867422/.\n",
      "    \n",
      "                        Alternatively, 'mutations' can be a string specifying a supported database, which will automatically download\n",
      "                        both the mutation database and corresponding reference sequence (if the 'sequences' is not a path).\n",
      "                        To see the supported combinations of mutations and sequences, either\n",
      "                        1) run `vk build --list_supported_databases` from the command line, or\n",
      "                        2) run varseek.build(list_supported_databases=True) in python\n",
      "    \n",
      "        # Additional parameters\n",
      "        - filters        (list) List of filters to apply to the mutations. See varseek filter documentation for more information.\n",
      "        - mode          (str) Mode to use for kb ref. Currently not implemented. Default: None.\n",
      "        - dlist         (str) Path to dlist fasta file or \"None\" (including the quotes). Default: False.\n",
      "        - config        (str) Path to config file. Default: None.\n",
      "        \n",
      "        # Optional output file paths: (only needed if changing/customizing file names or locations):\n",
      "        - out           (str) Output directory. Default: \".\".\n",
      "        - index_out (str) Path to output mcrs index file. Default: None.\n",
      "        - t2g_out (str) Path to output mcrs t2g file to be used in alignment. Default: None.\n",
      "        \n",
      "        # General arguments:\n",
      "        - download      (bool) If True, download prebuilt reference files. Default: False.\n",
      "        - dry_run       (bool) If True, print the commands that would be run without actually running them. Default: False.\n",
      "        - list_downloadable_references (bool) If True, list the available downloadable references. Default: False.\n",
      "        - minimum_info_columns (bool) If True, run vk info with minimum columns. Default: True.\n",
      "        - overwrite     (bool) If True, overwrite existing files. Default: False.\n",
      "        - threads       (int) Number of threads to use. Default: 2.\n",
      "        - verbose       (bool) If True, print progress messages. Default: True.\n",
      "        \n",
      "    ================================\n",
      "    \n",
      "    PARAMETERS FROM WRAPPED FUNCTIONS (note that any shared parameter names between functions are meant to have identical purposes)\n",
      "    \n",
      "    varseek build parameters\n",
      "        Takes in nucleotide sequences and variants (in standard mutation annotation - see below)\n",
      "        and returns sequences containing the variants and the surrounding local context, dubbed variant-containing reference sequences (VCRSs),\n",
      "        compatible with k-mer-based methods (i.e., kallisto | bustools) for variant detection.\n",
      "    \n",
      "        # Required input argument:\n",
      "        - sequences     (str) Path to the fasta file containing the sequences to have the mutations added, e.g., 'seqs.fa'.\n",
      "                        Sequence identifiers following the '>' character must correspond to the identifiers\n",
      "                        in the seq_ID column of 'mutations'.\n",
      "    \n",
      "                        Example:\n",
      "                        >seq1 (or ENSG00000106443)\n",
      "                        ACTGCGATAGACT\n",
      "                        >seq2\n",
      "                        AGATCGCTAG\n",
      "    \n",
      "                        Alternatively: Input sequence(s) as a string or a list of strings,\n",
      "                        e.g. 'AGCTAGCT' or ['ACTGCTAGCT', 'AGCTAGCT'].\n",
      "    \n",
      "                        NOTE: Only the letters until the first space or dot will be used as sequence identifiers\n",
      "                        - Version numbers of Ensembl IDs will be ignored.\n",
      "                        NOTE: When 'sequences' input is a genome, also see 'gtf' argument below.\n",
      "    \n",
      "                        Alternatively, if 'mutations' is a string specifying a supported database, \n",
      "                        sequences can be a string indicating the source upon which to apply the mutations.\n",
      "                        See below for supported databases and sequences options.\n",
      "                        To see the supported combinations of mutations and sequences, either\n",
      "                        1) run `vk build --list_supported_databases` from the command line, or\n",
      "                        2) run varseek.build(list_supported_databases=True) in python\n",
      "    \n",
      "        - mutations     (str or DataFrame object) Path to csv or tsv file (str) (e.g., 'mutations.csv'), or DataFrame (DataFrame object),\n",
      "                        containing information about the mutations in the following format:\n",
      "    \n",
      "                        | mutation         | mut_ID | seq_ID |\n",
      "                        | c.2C>T           | mut1   | seq1   | -> Apply mutation 1 to sequence 1\n",
      "                        | c.9_13inv        | mut2   | seq2   | -> Apply mutation 2 to sequence 2\n",
      "                        | c.9_13inv        | mut2   | seq3   | -> Apply mutation 2 to sequence 3\n",
      "                        | c.9_13delinsAAT  | mut3   | seq3   | -> Apply mutation 3 to sequence 3\n",
      "                        | ...              | ...    | ...    |\n",
      "    \n",
      "                        'mutation' = Column containing the mutations to be performed written in standard mutation annotation (see below)\n",
      "                        'seq_ID' = Column containing the identifiers of the sequences to be mutated (must correspond to the string following\n",
      "                        the > character in the 'sequences' fasta file; do NOT include spaces or dots)\n",
      "                        'mut_ID' = Column containing an identifier for each mutation (optional).\n",
      "    \n",
      "                        Alternatively: Input mutation(s) as a string or list, e.g., 'c.2C>T' or ['c.2C>T', 'c.1A>C'].\n",
      "                        If a list is provided, the number of mutations must equal the number of input sequences.\n",
      "    \n",
      "                        For more information on the standard mutation annotation, see https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1867422/.\n",
      "    \n",
      "                        Alternatively, 'mutations' can be a string specifying a supported database, which will automatically download\n",
      "                        both the mutation database and corresponding reference sequence (if the 'sequences' is not a path).\n",
      "                        To see the supported combinations of mutations and sequences, either\n",
      "                        1) run `vk build --list_supported_databases` from the command line, or\n",
      "                        2) run varseek.build(list_supported_databases=True) in python\n",
      "    \n",
      "        # Parameters affecting VCRS creation\n",
      "        - w                                  (int) Length of sequence windows flanking the variant. Default: 30.\n",
      "                                             If w > total length of the sequence, the entire sequence will be kept.\n",
      "        - k                                  (int) Length of the k-mers to be considered in remove_seqs_with_wt_kmers, and the default minimum value for the minimum sequence length (which can be changed with 'min_seq_len').\n",
      "                                             If using kallisto in a later workflow, then this should correspond to kallisto k.\n",
      "                                             Must be greater than the value passed in for w. Default: w+1.\n",
      "        - max_ambiguous                      (int) Maximum number of 'N' (or 'n') characters allowed in a VCRS. None means no 'N' filter will be applied. Default: 0.\n",
      "    \n",
      "        # Additional input files and associated parameters\n",
      "        - mut_column                         (str) Name of the column containing the variants to be introduced in 'mutations'. Default: 'mutation'.\n",
      "        - seq_id_column                      (str) Name of the column containing the IDs of the sequences to be mutated in 'mutations'. Default: 'seq_ID'.\n",
      "        - mut_id_column                      (str) Name of the column containing the IDs of each variant in 'mutations'. Optional. Default: use <seq_ID>_<mutation> for each row.\n",
      "        - gtf                                (str) Path to .gtf file. Only used in conjunction with the arguments `transcript_boundaries` and `identify_all_spliced_from_genome`. Default: None\n",
      "        - gtf_transcript_id_column           (str) Column name in the input 'mutations' file containing the transcript ID. \n",
      "                                             In this case, column seq_id_column should contain the chromosome number.\n",
      "                                             Required when 'gtf' is provided. Default: None\n",
      "        - transcript_boundaries              (True/False) Whether to use the transcript boundaries in the input 'gtf' file to define the boundaries of the VCRSs. Only used when the `sequences` and `mutations` information is in terms of the genome, and when `gtf` is specified. Default: False.\n",
      "        - identify_all_spliced_from_genome   (True/False) Whether to identify all spliced VCRSs from the genome. Default: False.\n",
      "        \n",
      "        # Output paths and associated parameters\n",
      "        - out                                (str) Path to default output directory to containing created files. Any individual output file path can be overriden if the specific file path is provided\n",
      "                                             as an argument. Default: \".\" (current directory).\n",
      "        - reference_out_dir                  (str) Path to reference file directory to be downloaded if 'mutations' is a supported database and the file corresponding to 'sequences' does not exist.\n",
      "                                             Default: <out>/reference.\n",
      "        - mcrs_fasta_out                     (str) Path to output fasta file containing the VCRSs. \n",
      "                                             If use_IDs=False, then the fasta headers will be the values in the column 'mut_ID' (semicolon-jooined if merge_identical=True).\n",
      "                                             Otherwise, if use_IDs=True (default), then the fasta headers will be of the form 'vcrs_<int>' where <int> is a unique integer. Default: \"<out>/mcrs.fa\"\n",
      "        - mutations_updated_csv_out          (str) Path to output csv file containing the updated DataFrame. Only valid if save_mutations_updated_csv=True. Default: \"<out>/mutation_metadata_df.csv\"\n",
      "        - id_to_header_csv_out               (str) File name of csv file containing the mapping of unique IDs to the original sequence headers if use_IDs=True. Default: \"<out>/id_to_header_mapping.csv\"\n",
      "        - mcrs_t2g_out                       (str) Path to output t2g file containing the transcript-to-gene mapping for the MCRSs. Used in kallisto | bustools workflow. Default: \"<out>/mcrs_t2g.txt\"\n",
      "        - wt_mcrs_fasta_out                  (str) Path to output fasta file containing the wildtype sequence counterparts of the mutation-containing reference sequences (MCRSs). Default: \"<out>/wt_mcrs.fa\"\n",
      "        - wt_mcrs_t2g_out                    (str) Path to output t2g file containing the transcript-to-gene mapping for the wildtype MCRSs. Default: \"<out>/wt_mcrs_t2g.txt\"\n",
      "        - removed_variants_text_out          (str) Path to output text file containing the removed variants. Default: \"<out>/removed_variants.txt\"\n",
      "        - filtering_report_text_out          (str) Path to output text file containing the filtering report. Default: \"<out>/filtering_report.txt\"\n",
      "    \n",
      "        # Returning and saving of optional output\n",
      "        - return_mutation_output             (True/False) Whether to return the mutation output saved in the fasta file. Default: False.\n",
      "        - save_mutations_updated_csv         (True/False) Whether to update the input 'mutations' DataFrame to include additional columns with the mutation type,\n",
      "                                             wildtype nucleotide sequence, and mutant nucleotide sequence (only valid if 'mutations' is a csv or tsv file). Default: False\n",
      "        - save_wt_mcrs_fasta_and_t2g         (True/False) Whether to create a fasta file containing the wildtype sequence counterparts of the mutation-containing reference sequences (MCRSs)\n",
      "                                             and the corresponding t2g. Default: False.\n",
      "        - save_removed_variants_text         (True/False) Whether to save a text file containing the removed variants. Default: True.\n",
      "        - save_filtering_report_text         (True/False) Whether to save a text file containing the filtering report. Default: True.\n",
      "        - store_full_sequences               (True/False) Whether to also include the complete wildtype and mutant sequences in the updated 'mutations' DataFrame (not just the sub-sequence with\n",
      "                                             w-length flanks). Only valid if save_mutations_updated_csv=True. Default: False\n",
      "        - translate                          (True/False) Add additional columns to the 'mutations' DataFrame containing the wildtype and mutant amino acid sequences.\n",
      "                                             Only valid if store_full_sequences=True. Default: False\n",
      "        - translate_start                    (int | str | None) The position in the input nucleotide sequence to start translating. If a string is provided, it should correspond\n",
      "                                             to a column name in 'mutations' containing the open reading frame start positions for each sequence/mutation.\n",
      "                                             Only valid if translate=True. Default: None (translate from the beginning of the sequence)\n",
      "        - translate_end                      (int | str | None) The position in the input nucleotide sequence to end translating. If a string is provided, it should correspond\n",
      "                                             to a column name in 'mutations' containing the open reading frame end positions for each sequence/mutation.\n",
      "                                             Only valid if translate=True. Default: None (translate from to the end of the sequence)\n",
      "    \n",
      "        # General arguments:\n",
      "        - dry_run                            (True/False) Whether to simulate the function call without executing it. Default: False.\n",
      "        - list_supported_databases           (True/False) Whether to print the supported databases and sequences. Default: False.\n",
      "        - overwrite                          (True/False) Whether to overwrite existing output files. Will return if any output file already exists. Default: False.\n",
      "        - verbose                            (True/False) Whether to print progress information. Default: True\n",
      "        \n",
      "        # # Hidden arguments (part of kwargs) - for niche use cases, specific databases, or debugging:\n",
      "        # # niche use cases\n",
      "        - insertion_size_limit               (int) Maximum number of nucleotides allowed in an insertion-type variant. Variants with insertions larger than this will be dropped.\n",
      "                                             Default: None (no insertion size limit will be applied)\n",
      "        - min_seq_len                        (int) Minimum length of the variant output sequence. Mutant sequences smaller than this will be dropped. None means no length filter will be applied. Default: k (from the \"k\" parameter)\n",
      "        - optimize_flanking_regions          (True/False) Whether to remove nucleotides from either end of the mutant sequence to ensure (when possible)\n",
      "                                             that the mutant sequence does not contain any (w+1)-mers (where a (w+1)-mer is a subsequence of length w+1, with w defined by the 'w' argument) also found in the wildtype/input sequence. Default: False\n",
      "        - remove_seqs_with_wt_kmers          (True/False) Removes output sequences where at least one k-mer is also present in the wildtype/input sequence in the same region.\n",
      "                                             If optimize_flanking_regions=True, only sequences for which a wildtype k-mer (with k defined by the 'k' argument) is still present after optimization will be removed.\n",
      "                                             Default: False\n",
      "        \n",
      "        - required_insertion_overlap_length  (int | str | None) Enforces the minimum number of bases included in the inserted region for all (w+1)-mers (where a (w+1)-mer is a subsequence of length w+1, with w defined by the 'w' argument),\n",
      "                                             or that all (w+1)-mers contain the entire inserted sequence (whatever is smaller). Only effective when optimize_flanking_regions is also True.\n",
      "                                             Default: None (Flank optimization occurs only until there is no shared k-mer in the VCRS and the reference sequence). If \"all\", then require the entire insertion and the following nucleotide (and filter out insertions of length >= 2*w)\n",
      "        - merge_identical                    (True/False) Whether to merge sequence-identical VCRSs in the output (identical VCRSs will be merged by concatenating the sequence\n",
      "                                             headers for all identical sequences with semicolons). Default: False\n",
      "        - vcrs_strandedness                  (True/False) Whether to consider the forward and reverse-complement mutant sequences as distinct if merging identical sequences. Only effective when merge_identical is also True. Default: False\n",
      "                                             Default: False (ie do not consider forward and reverse-complement sequences to be equivalent)\n",
      "        - use_IDs           (True/False) Whether to keep the original sequence headers in the output fasta file, or to replace them with unique IDs of the form 'vcrs_<int>.\n",
      "                                             If False, then an additional file at the path <id_to_header_csv_out> will be formed that maps sequence IDs from the fasta file to the <mut_id_column>. Default: True.\n",
      "        \n",
      "        # # specific databases\n",
      "        - cosmic_release                     (str) COSMIC release version to download. Default: \"100\".\n",
      "        - cosmic_grch                        (str) COSMIC genome reference version to download. Default: \"37\".\n",
      "        - cosmic_email                       (str) Email address for COSMIC download. Default: None.\n",
      "        - cosmic_password                    (str) Password for COSMIC download. Default: None.\n",
      "    \n",
      "        # # debugging\n",
      "        - save_files                         (True/False) Whether to save the output files. Default: True.\n",
      "    \n",
      "    \n",
      "        Saves mutated sequences in fasta format (or returns a list containing the mutated sequences if out=None).\n",
      "        \n",
      "    ================================\n",
      "    \n",
      "    \n",
      "        Takes in the input directory containing with the MCRS fasta file generated from varseek build, and returns a dataframe with additional columns containing information about the mutations.\n",
      "    \n",
      "        # Required input arguments:\n",
      "        - input_dir     (str) Path to the directory containing the input files. Corresponds to `out` in the varseek build function.\n",
      "    \n",
      "        # Additional Parameters\n",
      "        - columns_to_include                 (list or str) List of columns to include in the output dataframe. Default: (\"number_of_mutations_in_this_gene_total\", \"number_of_alignments_to_normal_human_reference\", \"pseudoaligned_to_human_reference_despite_not_truly_aligning\", \"longest_homopolymer_length\", \"triplet_complexity\"). See all possible values and their description by setting list_columns=True (python) or --list_columns (command line).\n",
      "        - k                                  (int) Length of the k-mers utilized by kallisto | bustools. Only used by the following columns: 'nearby_mutations', 'nearby_mutations_count', 'has_a_nearby_mutation', 'dlist', 'number_of_alignments_to_normal_human_reference', 'dlist_substring', 'number_of_substring_matches_to_normal_human_reference', 'pseudoaligned_to_human_reference', 'pseudoaligned_to_human_reference_despite_not_truly_aligning', 'number_of_kmers_with_overlap_to_other_mcrs_items_in_mcrs_reference', 'number_of_mcrs_items_with_overlapping_kmers_in_mcrs_reference', 'kmer_overlap_in_mcrs_reference'; and when make_kat_histogram==True. Default: 55.\n",
      "        - max_ambiguous_mcrs                 (int) Maximum number of 'N' characters allowed in the MCRS when considering alignment to the reference genome/transcriptome. Only used by the following columns: 'dlist', 'number_of_alignments_to_normal_human_reference', 'dlist_substring', 'number_of_substring_matches_to_normal_human_reference'. Default: 0.\n",
      "        - max_ambiguous_reference            (int) Maximum number of 'N' characters allowed in the aligned reference genome portion when considering alignment to the reference genome/transcriptome. Only used by the following columns: 'dlist', 'number_of_alignments_to_normal_human_reference', 'dlist_substring', 'number_of_substring_matches_to_normal_human_reference'. Default: 0.\n",
      "    \n",
      "        # Optional input file paths: (only needed if changing/customizing file names or locations):\n",
      "        - mcrs_fasta                         (str) Path to the MCRS fasta file generated from varseek build. Corresponds to `mcrs_fasta_out` in the varseek build function. Only needed if the original file was changed or renamed. Default: None (will find it in `input_dir`).\n",
      "        - mutations_updated_csv              (str) Path to the updated dataframe containing the MCRS headers and sequences. Corresponds to `mutations_updated_csv_out` in the varseek build function. Only needed if the original file was changed or renamed. Default: None (will find it in `input_dir` if it exists).\n",
      "        - id_to_header_csv                   (str) Path to the csv file containing the mapping of IDs to headers generated from varseek build corresponding to mcrs_fasta. Corresponds to `id_to_header_csv_out` in the varseek build function. Only needed if the original file was changed or renamed. Default: None (will find it in `input_dir` if it exists).\n",
      "        - gtf                                (str) Path to the GTF file containing the gene annotations for the reference genome. Corresponds to `gtf` in the varseek build function. Must align to genome coordinates used in the annotation of mutations. Only used by the following columns: 'distance_to_nearest_splice_junction'. Default: None.\n",
      "        - dlist_reference_genome_fasta       (str) Path to the reference genome fasta file for the d-list. Only used by the following columns: 'dlist', 'number_of_alignments_to_normal_human_reference', 'dlist_substring', 'number_of_substring_matches_to_normal_human_reference', 'pseudoaligned_to_human_reference', 'pseudoaligned_to_human_reference_despite_not_truly_aligning'. Default: \"T2T. (will automatically download the T2T reference genome files to `reference_out_dir`)\n",
      "        - dlist_reference_cdna_fasta         (str) Path to the reference cDNA fasta file for the d-list. Only used by the following columns: 'dlist', 'number_of_alignments_to_normal_human_reference', 'dlist_substring', 'number_of_substring_matches_to_normal_human_reference'. Default: \"T2T. (will automatically download the T2T reference genome files to `reference_out_dir`)\n",
      "        - dlist_reference_gtf                (str) Path to the GTF file containing the gene annotations for the reference genome. Only used by the following columns: 'pseudoaligned_to_human_reference', 'pseudoaligned_to_human_reference_despite_not_truly_aligning'. Default: \"T2T. (will automatically download the T2T reference genome files to `reference_out_dir`)\n",
      "    \n",
      "        # Column names in mutations_updated_csv:\n",
      "        - mcrs_id_column                     (str) Name of the column containing the MCRS IDs in `mutations_updated_csv`. Only used if `mutations_updated_csv` exists (i.e., was generated from varseek build). Default: 'mcrs_id'.\n",
      "        - mcrs_sequence_column               (str) Name of the column containing the MCRS sequences in `mutations_updated_csv`. Only used if `mutations_updated_csv` exists (i.e., was generated from varseek build). Default: 'mutant_sequence'.\n",
      "        - mcrs_source_column                 (str) Name of the column containing the source of the MCRS (cdna or genome) in `mutations_updated_csv`. Only used if `mutations_updated_csv` exists (i.e., was generated from varseek build). Default: 'mcrs_source'.\n",
      "        - mut_column                         (str) Name of the column containing the mutations in `mutations_updated_csv`. Only used if `mutations_updated_csv` exists (i.e., was generated from varseek build). Default: 'mutation'.\n",
      "        - seq_id_column                      (str) Name of the column containing the sequence IDs in `mutations_updated_csv`. Only used if `mutations_updated_csv` exists (i.e., was generated from varseek build). Default: 'seq_ID'.\n",
      "        - mutation_cdna_column               (str) Name of the column containing the cDNA mutations in `mutations_updated_csv`. Only used if `mutations_updated_csv` exists (i.e., was generated from varseek build) and contains information regarding both genome and cDNA notation (essential if running spliced + unspliced workflow, optional otherwise). Default: 'mutation'.\n",
      "        - seq_id_cdna_column                 (str) Name of the column containing the cDNA sequence IDs in `mutations_updated_csv`. Only used if `mutations_updated_csv` exists (i.e., was generated from varseek build) and contains information regarding both genome and cDNA notation (essential if running spliced + unspliced workflow, optional otherwise). Default: 'seq_ID'.\n",
      "        - mutation_genome_column             (str) Name of the column containing the genome mutations in `mutations_updated_csv`. Only used if `mutations_updated_csv` exists (i.e., was generated from varseek build) and contains information regarding both genome and cDNA notation (essential if running spliced + unspliced workflow, optional otherwise). Default: 'mutation_genome'.\n",
      "        - seq_id_genome_column               (str) Name of the column containing the genome sequence IDs in `mutations_updated_csv`. Only used if `mutations_updated_csv` exists (i.e., was generated from varseek build) and contains information regarding both genome and cDNA notation (essential if running spliced + unspliced workflow, optional otherwise). Default: 'chromosome'.\n",
      "    \n",
      "        # Output file paths:\n",
      "        - out                                (str) Path to the directory where the output files will be saved. Default: `input_dir`.\n",
      "        - reference_out_dir                      (str) Path to the directory where the reference files will be saved. Default: `out`.\n",
      "        - mutations_updated_vk_info_csv_out  (str) Path to the output csv file containing the updated dataframe with the additional columns. Default: `out`/mutation_metadata_df_updated_vk_info.csv.\n",
      "        - mutations_updated_exploded_vk_info_csv_out (str) Path to the output csv file containing the exploded dataframe with the additional columns. Default: `out`/mutation_metadata_df_updated_vk_info_exploded.csv.\n",
      "        - dlist_genome_fasta_out             (str) Path to the output fasta file containing the d-list sequences for the genome-based alignmed. Only used by the following columns: 'dlist', 'number_of_alignments_to_normal_human_reference', 'dlist_substring', 'number_of_substring_matches_to_normal_human_reference'. Default: `out`/dlist_genome.fa.\n",
      "        - dlist_cdna_fasta_out               (str) Path to the output fasta file containing the d-list sequences for the cDNA-based alignmed. Only used by the following columns: 'dlist', 'number_of_alignments_to_normal_human_reference', 'dlist_substring', 'number_of_substring_matches_to_normal_human_reference'.Default: `out`/dlist_cdna.fa.\n",
      "        - dlist_combined_fasta_out           (str) Path to the output fasta file containing the d-list sequences  combined genome-based and cDNA-based alignment. Only used by the following columns: 'dlist', 'number_of_alignments_to_normal_human_reference', 'dlist_substring', 'number_of_substring_matches_to_normal_human_reference'.Default: `out`/dlist.fa.\n",
      "    \n",
      "        # Returning and saving of optional output\n",
      "        - save_mutations_updated_exploded_vk_info_csv (bool) Whether to save the exploded dataframe. Default: False.\n",
      "        - make_pyfastx_summary_file          (bool) Whether to make a summary file of the MCRS fasta file using pyfastx. Default: False.\n",
      "        - make_kat_histogram                 (bool) Whether to make a histogram of the k-mer abundances using kat. Default: False.\n",
      "    \n",
      "        # General arguments:\n",
      "        - dry_run                            (bool) Whether to do a dry run (i.e., print the parameters and return without running the function). Default: False.\n",
      "        - list_columns                       (bool) Whether to list the possible values for `columns_to_include` and their descriptions. Default: False.\n",
      "        - overwrite                          (bool) Whether to overwrite the output files if they already exist. Default: False.\n",
      "        - threads                            (int) Number of threads to use for bowtie2 and bowtie2-build. Only used by the following columns: 'dlist', 'number_of_alignments_to_normal_human_reference', 'dlist_substring', 'number_of_substring_matches_to_normal_human_reference', 'pseudoaligned_to_human_reference', 'pseudoaligned_to_human_reference_despite_not_truly_aligning', 'entries_for_which_this_mcrs_is_substring', 'entries_for_which_this_mcrs_is_superstring', 'mcrs_is_substring', 'mcrs_is_superstring'; and when 'entries_for_which_this_mcrs_is_superstring', 'mcrs_is_substring', 'mcrs_is_superstring', make_kat_histogram==True. Default: 2.\n",
      "        - verbose                            (bool) Whether to print verbose output. Default: True.\n",
      "    \n",
      "        # Hidden arguments (part of kwargs):\n",
      "        - w                                  (int) Maximum length of the MCRS flanking regions. Must be an integer between [1, k-1]. Only utilized for the column 'cdna_and_genome_same'. Corresponds to `w` in the varseek build function. Default: 54.\n",
      "        - bowtie_path                        (str) Path to the directory containing the bowtie2 and bowtie2-build executables. Default: None.\n",
      "        - vcrs_strandedness                  (bool) Whether to consider MCRSs as stranded when aligning to the human reference and comparing MCRS k-mers to each other. vcrs_strandedness True corresponds to treating forward and reverse-complement as distinct; False corresponds to treating them as the same. Corresponds to `vcrs_strandedness` in the varseek build function. Only used by the following columns: 'dlist', 'number_of_alignments_to_normal_human_reference', 'dlist_substring', 'number_of_substring_matches_to_normal_human_reference', 'pseudoaligned_to_human_reference', 'pseudoaligned_to_human_reference_despite_not_truly_aligning', 'number_of_kmers_with_overlap_to_other_mcrs_items_in_mcrs_reference', 'number_of_mcrs_items_with_overlapping_kmers_in_mcrs_reference', 'kmer_overlap_in_mcrs_reference'; and if make_kat_histogram==True. Default: False.\n",
      "        - near_splice_junction_threshold     (int) Maximum distance from a splice junction to be considered \"near\" a splice junction. Only utilized for the column 'distance_to_nearest_splice_junction'. Default: 10.\n",
      "        - reference_cdna_fasta               (str) Path to the reference cDNA fasta file. Only utilized for the column 'cdna_and_genome_same'. Default: None.\n",
      "        - reference_genome_fasta             (str) Path to the reference genome fasta file. Only utilized for the column 'cdna_and_genome_same'. Default: None.\n",
      "        - mutations_csv                      (str) Path to the mutations csv file. Only utilized for the column 'cdna_and_genome_same'. Corresponds to `mutations` in the varseek build function. Default: None.\n",
      "        \n",
      "    ================================\n",
      "    \n",
      "    \n",
      "        Filter mutations based on the provided filters and save the filtered mutations to a fasta file.\n",
      "    \n",
      "        # Required input arguments:\n",
      "        - input_dir     (str) Path to the directory containing the input files. Corresponds to `out` in the varseek info function.\n",
      "        - filters       (str or list) String or list of filters to apply. If a string, it should be a path to a txt file containing the filters. If a list, it should be a list of strings in the format COLUMNNAME-RULE=VALUE. See documentation for details, or run vk.filter(list_filter_rules=True).\n",
      "    \n",
      "        # Optional input arguments:\n",
      "        - mutations_updated_vk_info_csv                (str) Path to the updated dataframe containing the MCRS headers and sequences. Corresponds to `mutations_updated_csv_out` in the varseek build function. Only needed if the original file was changed or renamed. Default: None (will find it in `input_dir` if it exists).\n",
      "        - mutations_updated_exploded_vk_info_csv       (str) Path to the updated exploded dataframe containing the MCRS headers and sequences. Corresponds to `mutations_updated_exploded_csv_out` in the varseek build function. Only needed if the original file was changed or renamed. Default: None (will find it in `input_dir` if it exists).\n",
      "        - id_to_header_csv                             (str) Path to the csv file containing the mapping of IDs to headers generated from varseek build corresponding to mcrs_fasta. Corresponds to `id_to_header_csv_out` in the varseek build function. Only needed if the original file was changed or renamed. Default: None (will find it in `input_dir` if it exists).\n",
      "        - dlist_fasta                                  (str) Path to the dlist fasta file. Default: None (will find it in `input_dir` if it exists).\n",
      "    \n",
      "        # Optional output file paths: (only needed if changing/customizing file names or locations):\n",
      "        - mutations_updated_filtered_csv_out           (str) Path to the filtered mutation metadata dataframe. Default: None (will be saved in `out`).\n",
      "        - mutations_updated_exploded_filtered_csv_out  (str) Path to the filtered exploded mutation metadata dataframe. Default: None (will be saved in `out`).\n",
      "        - id_to_header_filtered_csv_out                (str) Path to the filtered id to header csv. Default: None (will be saved in `out`).\n",
      "        - dlist_filtered_fasta_out                     (str) Path to the filtered dlist fasta file. Default: None (will be saved in `out`).\n",
      "        - mcrs_filtered_fasta_out                      (str) Path to the filtered mcrs fasta file. Default: None (will be saved in `out`).\n",
      "        - mcrs_t2g_filtered_out                        (str) Path to the filtered t2g file. Default: None (will be saved in `out`).\n",
      "        - wt_mcrs_filtered_fasta_out                   (str) Path to the filtered wt mcrs fasta file. Default: None (will be saved in `out`).\n",
      "        - wt_mcrs_t2g_filtered_out                     (str) Path to the filtered t2g file for wt mcrs fasta. Default: None (will be saved in `out`).\n",
      "    \n",
      "        # Returning and saving of optional output\n",
      "        - save_wt_mcrs_fasta_and_t2g                   (bool) If True, save the filtered wt mcrs fasta and t2g files. Default: False.\n",
      "        - save_mutations_updated_filtered_csvs         (bool) If True, save the filtered mutation metadata dataframe. Default: False.\n",
      "        - return_mutations_updated_filtered_csv_df     (bool) If True, return the filtered mutation metadata dataframe. Default: False.\n",
      "    \n",
      "        # General arguments:\n",
      "        - dry_run                                      (bool) If True, print the parameters and exit without running the function. Default: False.\n",
      "        - list_filter_rules                            (bool) If True, print the available filter rules and exit without running the function. Default: False.\n",
      "        - overwrite                                    (bool) If True, overwrite the output files if they already exist. Default: False.\n",
      "        - verbose                                      (bool) If True, print progress messages. Default: True.\n",
      "    \n",
      "        # Hidden arguments:\n",
      "        filter_all_dlists (bool) If True, filter all dlists. Default: False.\n",
      "        dlist_genome_fasta (str) Path to the genome dlist fasta file. Default: None.\n",
      "        dlist_cdna_fasta (str) Path to the cDNA dlist fasta file. Default: None.\n",
      "        dlist_genome_filtered_fasta_out (str) Path to the filtered genome dlist fasta file. Default: None.\n",
      "        dlist_cdna_filtered_fasta_out (str) Path to the filtered cDNA dlist fasta file. Default: None.\n",
      "        save_mcrs_filtered_fasta_and_t2g (bool) If True, save the filtered mcrs fasta and t2g files. Default: True.\n",
      "        \n",
      "    ================================\n",
      "    \n",
      "    kb ref parameters (see `kb ref --help` for more information on kb ref arguments)\n",
      "\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "import varseek as vk\n",
    "print(help(vk.ref))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "âš ï¸ seqtk is not configured properly.\n",
      "ðŸ”„ Attempting to install seqtk...\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[14], line 45\u001b[0m\n\u001b[1;32m     42\u001b[0m         \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mâŒ Installation failed. Install manually.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     44\u001b[0m \u001b[38;5;66;03m# Run check\u001b[39;00m\n\u001b[0;32m---> 45\u001b[0m \u001b[43mcheck_and_install_seqtk\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[14], line 18\u001b[0m, in \u001b[0;36mcheck_and_install_seqtk\u001b[0;34m()\u001b[0m\n\u001b[1;32m     15\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mâŒ seqtk is not installed.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     17\u001b[0m \u001b[38;5;66;03m# Attempt installation\u001b[39;00m\n\u001b[0;32m---> 18\u001b[0m \u001b[43minstall_seqtk\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     19\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;01mFalse\u001b[39;00m\n",
      "Cell \u001b[0;32mIn[14], line 39\u001b[0m, in \u001b[0;36minstall_seqtk\u001b[0;34m()\u001b[0m\n\u001b[1;32m     36\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m\n\u001b[1;32m     38\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m---> 39\u001b[0m     \u001b[43msubprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrun\u001b[49m\u001b[43m(\u001b[49m\u001b[43minstall_cmd\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcheck\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m)\u001b[49m\n\u001b[1;32m     40\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mâœ… seqtk installed successfully!\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     41\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m subprocess\u001b[38;5;241m.\u001b[39mCalledProcessError:\n",
      "File \u001b[0;32m~/miniconda3/envs/varseek/lib/python3.10/subprocess.py:505\u001b[0m, in \u001b[0;36mrun\u001b[0;34m(input, capture_output, timeout, check, *popenargs, **kwargs)\u001b[0m\n\u001b[1;32m    503\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Popen(\u001b[38;5;241m*\u001b[39mpopenargs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs) \u001b[38;5;28;01mas\u001b[39;00m process:\n\u001b[1;32m    504\u001b[0m     \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 505\u001b[0m         stdout, stderr \u001b[38;5;241m=\u001b[39m \u001b[43mprocess\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcommunicate\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43minput\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    506\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m TimeoutExpired \u001b[38;5;28;01mas\u001b[39;00m exc:\n\u001b[1;32m    507\u001b[0m         process\u001b[38;5;241m.\u001b[39mkill()\n",
      "File \u001b[0;32m~/miniconda3/envs/varseek/lib/python3.10/subprocess.py:1146\u001b[0m, in \u001b[0;36mPopen.communicate\u001b[0;34m(self, input, timeout)\u001b[0m\n\u001b[1;32m   1144\u001b[0m         stderr \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mread()\n\u001b[1;32m   1145\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mstderr\u001b[38;5;241m.\u001b[39mclose()\n\u001b[0;32m-> 1146\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwait\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1147\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1148\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/varseek/lib/python3.10/subprocess.py:1209\u001b[0m, in \u001b[0;36mPopen.wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1207\u001b[0m     endtime \u001b[38;5;241m=\u001b[39m _time() \u001b[38;5;241m+\u001b[39m timeout\n\u001b[1;32m   1208\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1209\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1210\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyboardInterrupt\u001b[39;00m:\n\u001b[1;32m   1211\u001b[0m     \u001b[38;5;66;03m# https://bugs.python.org/issue25942\u001b[39;00m\n\u001b[1;32m   1212\u001b[0m     \u001b[38;5;66;03m# The first keyboard interrupt waits briefly for the child to\u001b[39;00m\n\u001b[1;32m   1213\u001b[0m     \u001b[38;5;66;03m# exit under the common assumption that it also received the ^C\u001b[39;00m\n\u001b[1;32m   1214\u001b[0m     \u001b[38;5;66;03m# generated SIGINT and will exit rapidly.\u001b[39;00m\n\u001b[1;32m   1215\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/varseek/lib/python3.10/subprocess.py:1959\u001b[0m, in \u001b[0;36mPopen._wait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m   1957\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mreturncode \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m   1958\u001b[0m     \u001b[38;5;28;01mbreak\u001b[39;00m  \u001b[38;5;66;03m# Another thread waited.\u001b[39;00m\n\u001b[0;32m-> 1959\u001b[0m (pid, sts) \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_try_wait\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1960\u001b[0m \u001b[38;5;66;03m# Check the pid and loop as waitpid has been known to\u001b[39;00m\n\u001b[1;32m   1961\u001b[0m \u001b[38;5;66;03m# return 0 even without WNOHANG in odd situations.\u001b[39;00m\n\u001b[1;32m   1962\u001b[0m \u001b[38;5;66;03m# http://bugs.python.org/issue14396.\u001b[39;00m\n\u001b[1;32m   1963\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m pid \u001b[38;5;241m==\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid:\n",
      "File \u001b[0;32m~/miniconda3/envs/varseek/lib/python3.10/subprocess.py:1917\u001b[0m, in \u001b[0;36mPopen._try_wait\u001b[0;34m(self, wait_flags)\u001b[0m\n\u001b[1;32m   1915\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"All callers to this function MUST hold self._waitpid_lock.\"\"\"\u001b[39;00m\n\u001b[1;32m   1916\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1917\u001b[0m     (pid, sts) \u001b[38;5;241m=\u001b[39m \u001b[43mos\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwaitpid\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpid\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mwait_flags\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1918\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mChildProcessError\u001b[39;00m:\n\u001b[1;32m   1919\u001b[0m     \u001b[38;5;66;03m# This happens if SIGCLD is set to be ignored or waiting\u001b[39;00m\n\u001b[1;32m   1920\u001b[0m     \u001b[38;5;66;03m# for child processes has otherwise been disabled for our\u001b[39;00m\n\u001b[1;32m   1921\u001b[0m     \u001b[38;5;66;03m# process.  This child is dead, we can't get the status.\u001b[39;00m\n\u001b[1;32m   1922\u001b[0m     pid \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mpid\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "==> Auto-updating Homebrew...\n",
      "Adjust how often this is run with HOMEBREW_AUTO_UPDATE_SECS or disable with\n",
      "HOMEBREW_NO_AUTO_UPDATE. Hide these hints with HOMEBREW_NO_ENV_HINTS (see `man brew`).\n",
      "==> Downloading https://ghcr.io/v2/homebrew/portable-ruby/portable-ruby/blobs/sha256:4ffc8607e08e9bd536f1df71643b2ecb4cea1a15be9226f297008bc34d0bc8e2\n",
      "######################################################################## 100.0%\n",
      "==> Pouring portable-ruby-3.3.7.el_capitan.bottle.tar.gz\n",
      "==> Auto-updated Homebrew!\n",
      "==> Updated Homebrew from 4.4.6 (c81cd0b929) to 4.4.19 (4ee6e96bdf).\n",
      "Updated 3 taps (brewsci/bio, homebrew/core and homebrew/cask).\n",
      "==> New Formulae\n",
      "acme.sh\n",
      "acronym\n",
      "aliae\n",
      "alive2\n",
      "ampl-asl\n",
      "ansible@10\n",
      "asciigen\n",
      "atac\n",
      "aws-c-auth\n",
      "aws-c-cal\n",
      "aws-c-common\n",
      "aws-c-compression\n",
      "aws-c-event-stream\n",
      "aws-c-http\n",
      "aws-c-io\n",
      "aws-c-mqtt\n",
      "aws-c-s3\n",
      "aws-c-sdkutils\n",
      "aws-checksums\n",
      "aws-crt-cpp\n",
      "azure-core-cpp\n",
      "azure-storage-blobs-cpp\n",
      "azure-storage-common-cpp\n",
      "azurehound\n",
      "babelfish\n",
      "beanquery\n",
      "bender\n",
      "binocle\n",
      "bold\n",
      "brewsci/bio/boost-python3@1.87\n",
      "bruno-cli\n",
      "c2patool\n",
      "cargo-cache\n",
      "cargo-chef\n",
      "cargo-cyclonedx\n",
      "cargo-dist\n",
      "cargo-expand\n",
      "cargo-flamegraph\n",
      "cargo-hack\n",
      "cargo-msrv\n",
      "cargo-public-api\n",
      "cargo-run-bin\n",
      "cargo-shuttle\n",
      "cargo-update\n",
      "cdncheck\n",
      "chaos-client\n",
      "chiko\n",
      "cloudquery\n",
      "cmake-lint\n",
      "cobo-cli\n",
      "cobra-cli\n",
      "codeberg-cli\n",
      "codesnap\n",
      "codevis\n",
      "cpp-peglib\n",
      "crates-tui\n",
      "diary\n",
      "directx-headers\n",
      "dpic\n",
      "draft\n",
      "dragonbox\n",
      "dry\n",
      "ducker\n",
      "dum\n",
      "eslint_d\n",
      "eventpp\n",
      "fairy-stockfish\n",
      "fastbuild\n",
      "fish-lsp\n",
      "fltk@1.3\n",
      "funzzy\n",
      "fzf-make\n",
      "gci\n",
      "gh-ost\n",
      "ghc@9.10\n",
      "git-ignore\n",
      "glaze\n",
      "glibc@2.17\n",
      "go-blueprint\n",
      "go-parquet-tools\n",
      "gopeed\n",
      "graphql-inspector\n",
      "gsan\n",
      "guile-fibers\n",
      "gurk\n",
      "harlequin\n",
      "harper\n",
      "havn\n",
      "heroku\n",
      "hevi\n",
      "hgrep\n",
      "html2markdown\n",
      "humanlog\n",
      "idsgrep\n",
      "imgp\n",
      "ipatool\n",
      "json2hcl\n",
      "jwt-cli\n",
      "kfr\n",
      "kool\n",
      "kraftkit\n",
      "kube-bench\n",
      "kubeone\n",
      "kubernetes-cli@1.31\n",
      "kubetrim\n",
      "languagetool-rust\n",
      "leetcode-cli\n",
      "legba\n",
      "libcotp\n",
      "libcutl\n",
      "libesedb\n",
      "libformfactor\n",
      "libgit2@1.8\n",
      "libheinz\n",
      "libkeyfinder\n",
      "librime\n",
      "libsidplayfp\n",
      "libsvgtiny\n",
      "libultrahdr\n",
      "libunicode\n",
      "libxsd-frontend\n",
      "limbo\n",
      "lla\n",
      "lol-html\n",
      "ltex-ls-plus\n",
      "ludusavi\n",
      "macmon\n",
      "mago\n",
      "malcontent\n",
      "marisa\n",
      "mdz\n",
      "meli\n",
      "minio-warp\n",
      "mmdbctl\n",
      "monocle\n",
      "netatalk\n",
      "neverest\n",
      "nrm\n",
      "odbc2parquet\n",
      "openapi-tui\n",
      "oranda\n",
      "parallel-disk-usage\n",
      "parca\n",
      "parlay\n",
      "percona-server@8.0\n",
      "percona-xtrabackup@8.0\n",
      "periphery\n",
      "perl-dbd-mysql\n",
      "php@8.3\n",
      "pie\n",
      "pinact\n",
      "pipelight\n",
      "pistache\n",
      "pixi-pack\n",
      "pkgx\n",
      "planus\n",
      "pnpm@9\n",
      "poselib\n",
      "posting\n",
      "precious\n",
      "prism-cli\n",
      "protoc-gen-grpc-java\n",
      "proxyfor\n",
      "pscale\n",
      "qbec\n",
      "rainfrog\n",
      "rapidyaml\n",
      "ratify\n",
      "redka\n",
      "redocly-cli\n",
      "regal\n",
      "render\n",
      "replxx\n",
      "repren\n",
      "reveal-md\n",
      "rmtrash\n",
      "rogcat\n",
      "rshijack\n",
      "ruby@3.3\n",
      "runitor\n",
      "s5cmd\n",
      "scooter\n",
      "sdl3\n",
      "sendme\n",
      "sentry-cli\n",
      "sequin\n",
      "serie\n",
      "sesh\n",
      "sfml@2\n",
      "sfsexp\n",
      "shepherd\n",
      "showcert\n",
      "sixtunnel\n",
      "slepc\n",
      "slepc-complex\n",
      "slumber\n",
      "spytrap-adb\n",
      "sqlitecpp\n",
      "sshx\n",
      "starlark-rust\n",
      "static-web-server\n",
      "streamrip\n",
      "superhtml\n",
      "swc\n",
      "swctl\n",
      "synchrony\n",
      "tabulate\n",
      "tach\n",
      "talosctl\n",
      "television\n",
      "templ\n",
      "tenere\n",
      "termbg\n",
      "tfocus\n",
      "the-way\n",
      "tiny\n",
      "tinysparql\n",
      "tomcat@10\n",
      "topiary\n",
      "tuisky\n",
      "typeshare\n",
      "tzdb\n",
      "ubi\n",
      "umka-lang\n",
      "vals\n",
      "versitygw\n",
      "vgt\n",
      "victorialogs\n",
      "vo-amrwbenc\n",
      "vue-language-server\n",
      "wails\n",
      "watcher\n",
      "wcstools\n",
      "weaviate\n",
      "weaviate-cli\n",
      "wgo\n",
      "xc\n",
      "xcresultparser\n",
      "xmq\n",
      "xtermcontrol\n",
      "yozefu\n",
      "zigup\n",
      "zsh-f-sy-h\n",
      "==> New Casks\n",
      "alcove\n",
      "amiberry\n",
      "aw-edid-editor\n",
      "ayugram\n",
      "bananas\n",
      "beaver-notes\n",
      "bindiff\n",
      "bookmacster@beta\n",
      "clipbook\n",
      "cpu-info\n",
      "dehesselle-meld\n",
      "display-pilot\n",
      "dockfix\n",
      "dorico\n",
      "fastdmg\n",
      "flashspace\n",
      "flowvision\n",
      "font-42dot-sans\n",
      "font-agu-display\n",
      "font-atkinson-hyperlegible-mono\n",
      "font-atkinson-hyperlegible-next\n",
      "font-badeen-display\n",
      "font-bitcount\n",
      "font-bitcount-prop-single-ink\n",
      "font-bitcount-single-ink\n",
      "font-brass-mono\n",
      "font-bravura\n",
      "font-cica-without-emoji\n",
      "font-fairfax-hd\n",
      "font-gd-highwaygothicja\n",
      "font-ibm-plex-sans-sc\n",
      "font-kanchenjunga\n",
      "font-kreative-square\n",
      "font-liter\n",
      "font-monomakh\n",
      "font-moralerspace\n",
      "font-moralerspace-hw\n",
      "font-moralerspace-hw-jpdoc\n",
      "font-moralerspace-hw-nf\n",
      "font-moralerspace-jpdoc\n",
      "font-moralerspace-nf\n",
      "font-nanum-square\n",
      "font-nanum-square-neo\n",
      "font-nanum-square-round\n",
      "font-pixel-code\n",
      "font-playwrite-ar-guides\n",
      "font-playwrite-at-guides\n",
      "font-playwrite-au-nsw-guides\n",
      "font-playwrite-au-qld-guides\n",
      "font-playwrite-au-sa-guides\n",
      "font-playwrite-au-tas-guides\n",
      "font-playwrite-au-vic-guides\n",
      "font-playwrite-be-vlg-guides\n",
      "font-playwrite-be-wal-guides\n",
      "font-playwrite-br-guides\n",
      "font-playwrite-ca-guides\n",
      "font-playwrite-cl-guides\n",
      "font-playwrite-co-guides\n",
      "font-playwrite-cu-guides\n",
      "font-playwrite-cz-guides\n",
      "font-playwrite-de-grund-guides\n",
      "font-playwrite-de-la-guides\n",
      "font-playwrite-de-sas-guides\n",
      "font-playwrite-de-va-guides\n",
      "font-playwrite-dk-loopet-guides\n",
      "font-playwrite-dk-uloopet-guides\n",
      "font-playwrite-es-deco-guides\n",
      "font-playwrite-es-guides\n",
      "font-playwrite-fr-moderne-guides\n",
      "font-playwrite-fr-trad-guides\n",
      "font-playwrite-gb-j-guides\n",
      "font-playwrite-gb-s-guides\n",
      "font-playwrite-hr-guides\n",
      "font-playwrite-hr-lijeva-guides\n",
      "font-playwrite-hu-guides\n",
      "font-playwrite-id-guides\n",
      "font-playwrite-ie-guides\n",
      "font-playwrite-in-guides\n",
      "font-playwrite-is-guides\n",
      "font-playwrite-it-moderna-guides\n",
      "font-playwrite-it-trad-guides\n",
      "font-playwrite-mx-guides\n",
      "font-playwrite-ng-modern-guides\n",
      "font-playwrite-nl-guides\n",
      "font-playwrite-no-guides\n",
      "font-playwrite-nz-guides\n",
      "font-playwrite-pe-guides\n",
      "font-playwrite-pl-guides\n",
      "font-playwrite-pt-guides\n",
      "font-playwrite-ro-guides\n",
      "font-playwrite-sk-guides\n",
      "font-playwrite-tz-guides\n",
      "font-playwrite-us-modern-guides\n",
      "font-playwrite-us-trad-guides\n",
      "font-playwrite-vn-guides\n",
      "font-playwrite-za-guides\n",
      "font-pochaevsk\n",
      "font-ponomar\n",
      "font-shafarik\n",
      "font-sketchybar-app-font\n",
      "font-tagesschrift\n",
      "font-tamzen\n",
      "font-triodion\n",
      "font-udev-gothic-hs\n",
      "fontcreator\n",
      "ghostty\n",
      "ghostty@tip\n",
      "green-go-control\n",
      "halion-sonic\n",
      "helio\n",
      "helpwire-operator\n",
      "ishare\n",
      "istherenet\n",
      "jottacloud\n",
      "linearmouse@beta\n",
      "lyric-fever\n",
      "macai\n",
      "meridiem\n",
      "mihomo-party\n",
      "mindmaster-cn\n",
      "mixing-station\n",
      "mouseless\n",
      "muteme\n",
      "nexonplug\n",
      "nvidia-nsight-systems\n",
      "octarine\n",
      "ontime\n",
      "openchrom\n",
      "openhv\n",
      "radiola\n",
      "readest\n",
      "restic-browser\n",
      "rivet\n",
      "rode-virtual-channels\n",
      "satdump\n",
      "satyrn\n",
      "serial-studio\n",
      "shortwave\n",
      "soundsource@test\n",
      "spectrolite\n",
      "steinberg-download-assistant\n",
      "steinberg-library-manager\n",
      "steinberg-mediabay\n",
      "stratoshark\n",
      "stratoshark@dev\n",
      "superwhisper\n",
      "supremo\n",
      "sys-pc-tool\n",
      "tabtab\n",
      "taskbar\n",
      "tdr-molotok\n",
      "tdr-prism\n",
      "termora\n",
      "testfully\n",
      "tm-error-logger\n",
      "tpvirtual\n",
      "valhalla-freq-echo\n",
      "valhalla-space-modulator\n",
      "valhalla-supermassive\n",
      "website-audit\n",
      "widgettoggler\n",
      "windowkeys\n",
      "windsurf\n",
      "wiso-steuer-2025\n",
      "wizcli\n",
      "zen-browser@twilight\n",
      "\n",
      "You have 124 outdated formulae installed.\n",
      "\n",
      "\n",
      "The 4.4.19 changelog can be found at:\n",
      "  https://github.com/Homebrew/brew/releases/tag/4.4.19\n"
     ]
    }
   ],
   "source": [
    "import subprocess\n",
    "import sys\n",
    "\n",
    "def check_and_install_seqtk():\n",
    "    try:\n",
    "        # Try running `seqtk` to check if it's installed\n",
    "        result = subprocess.run([\"seqtk\"], stdout=subprocess.PIPE, stderr=subprocess.PIPE)\n",
    "        \n",
    "        if result.returncode == 0:\n",
    "            print(\"âœ… seqtk is installed and working.\")\n",
    "            return True\n",
    "        else:\n",
    "            print(\"âš ï¸ seqtk is not configured properly.\")\n",
    "    except FileNotFoundError:\n",
    "        print(\"âŒ seqtk is not installed.\")\n",
    "\n",
    "    # Attempt installation\n",
    "    install_seqtk()\n",
    "    return False\n",
    "\n",
    "def install_seqtk():\n",
    "    print(\"ðŸ”„ Attempting to install seqtk...\")\n",
    "\n",
    "    if sys.platform == \"darwin\":  # macOS\n",
    "        install_cmd = [\"brew\", \"install\", \"seqtk\"]\n",
    "    elif shutil.which(\"mamba\"):\n",
    "        install_cmd = [\"mamba\", \"install\", \"-y\", \"seqtk\", \"-c\", \"bioconda\"]\n",
    "    elif shutil.which(\"conda\"):\n",
    "        install_cmd = [\"conda\", \"install\", \"-y\", \"seqtk\", \"-c\", \"bioconda\"]\n",
    "    elif shutil.which(\"apt\"):\n",
    "        install_cmd = [\"sudo\", \"apt\", \"install\", \"-y\", \"seqtk\"]\n",
    "    elif shutil.which(\"yum\"):\n",
    "        install_cmd = [\"sudo\", \"yum\", \"install\", \"-y\", \"seqtk\"]\n",
    "    else:\n",
    "        print(\"âŒ No supported package manager found. Install seqtk manually.\")\n",
    "        return\n",
    "\n",
    "    try:\n",
    "        subprocess.run(install_cmd, check=True)\n",
    "        print(\"âœ… seqtk installed successfully!\")\n",
    "    except subprocess.CalledProcessError:\n",
    "        print(\"âŒ Installation failed. Install manually.\")\n",
    "\n",
    "# Run check\n",
    "check_and_install_seqtk()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import inspect\n",
    "import re\n",
    "import varseek as vk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[W::vcf_parse] Contig '19' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig '20' is not defined in the header. (Quick workaround: index the file with tabix.)\n",
      "[W::vcf_parse] Contig 'X' is not defined in the header. (Quick workaround: index the file with tabix.)\n"
     ]
    }
   ],
   "source": [
    "from varseek.utils import vcf_to_dataframe\n",
    "vcf_file_path = \"/Users/joeyrich/Downloads/sample.vcf\"\n",
    "vcf_df = vcf_to_dataframe(vcf_file_path, additional_columns = True, explode_alt = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>mut_column</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>111</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>g.111A&gt;C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>112</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>g.112A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>14370</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>g.14370G&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>17330</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>g.17330T&gt;A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1110696</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>g.1110696A&gt;G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1110696</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>g.1110696A&gt;T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1230237</td>\n",
       "      <td>T</td>\n",
       "      <td>None</td>\n",
       "      <td>g.UNKNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1234567</td>\n",
       "      <td>G</td>\n",
       "      <td>GA</td>\n",
       "      <td>g.1234567_1234568insA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1234567</td>\n",
       "      <td>G</td>\n",
       "      <td>GAC</td>\n",
       "      <td>g.1234567_1234568insAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1235237</td>\n",
       "      <td>T</td>\n",
       "      <td>None</td>\n",
       "      <td>g.UNKNOWN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>10</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>g.11del</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>10</td>\n",
       "      <td>AC</td>\n",
       "      <td>ATG</td>\n",
       "      <td>g.10_11delinsATG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        POS REF   ALT              mut_column\n",
       "0       111   A     C                g.111A>C\n",
       "1       112   A     G                g.112A>G\n",
       "2     14370   G     A              g.14370G>A\n",
       "3     17330   T     A              g.17330T>A\n",
       "4   1110696   A     G            g.1110696A>G\n",
       "5   1110696   A     T            g.1110696A>T\n",
       "6   1230237   T  None               g.UNKNOWN\n",
       "7   1234567   G    GA   g.1234567_1234568insA\n",
       "8   1234567   G   GAC  g.1234567_1234568insAC\n",
       "9   1235237   T  None               g.UNKNOWN\n",
       "10       10  AC     A                 g.11del\n",
       "11       10  AC   ATG        g.10_11delinsATG"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_df[[\"POS\", \"REF\", \"ALT\", var_column]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Default value of k: 54\n"
     ]
    }
   ],
   "source": [
    "import inspect\n",
    "from varseek.varseek_build import build  # Import the function\n",
    "\n",
    "# Get function signature\n",
    "signature = inspect.signature(build)\n",
    "\n",
    "# Retrieve the default value for 'k'\n",
    "k_default = signature.parameters[\"w\"].default\n",
    "\n",
    "print(f\"Default value of k: {k_default}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Toy DataFrame:\n",
      "   semicolon_count\n",
      "0                0\n",
      "1                0\n",
      "2                1\n",
      "3                2\n",
      "4                3\n",
      "\n",
      "Total Sum: 9\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Create a toy DataFrame\n",
    "data = {\"semicolon_count\": [0, 0, 1, 2, 3]}\n",
    "mutation_metadata_df = pd.DataFrame(data)\n",
    "\n",
    "# Compute the sum as per your requirement\n",
    "total_sum = (mutation_metadata_df.loc[mutation_metadata_df[\"semicolon_count\"] > 0, \"semicolon_count\"] + 1).sum()\n",
    "\n",
    "print(\"Toy DataFrame:\")\n",
    "print(mutation_metadata_df)\n",
    "print(\"\\nTotal Sum:\", total_sum)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "'Signature' object is not iterable",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m key \u001b[38;5;129;01min\u001b[39;00m signature:\n\u001b[1;32m      2\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mParameter: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mkey\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m      3\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mDefault: \u001b[39m\u001b[38;5;132;01m{\u001b[39;00msignature\u001b[38;5;241m.\u001b[39mparameters[key]\u001b[38;5;241m.\u001b[39mdefault\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mTypeError\u001b[0m: 'Signature' object is not iterable"
     ]
    }
   ],
   "source": [
    "for key in signature:\n",
    "    print(f\"Parameter: {key}\")\n",
    "    print(f\"Default: {signature.parameters[key].default}\")\n",
    "    print(f\"Type: {signature.parameters[key].annotation}\")\n",
    "    print(f\"Description: {signature.parameters[key].annotation.__doc__}\")\n",
    "    print()\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "explicit_parameters_vk_build = vk.utils.get_set_of_parameters_from_function_signature(vk.varseek_build.build)\n",
    "allowable_kwargs_vk_build = vk.utils.get_set_of_allowable_kwargs(vk.varseek_build.build)\n",
    "explicit_parameters_vk_build = explicit_parameters_vk_build.union(allowable_kwargs_vk_build)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{1, 2, 3, 4, 5}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "if not isinstance(value, int) and not (isinstance(value, str) and value.isdigit()):\n",
    "    print(\"Value is not an integer or a string that can be converted to an integer\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'cosmic_email',\n",
       " 'cosmic_grch',\n",
       " 'cosmic_password',\n",
       " 'cosmic_release',\n",
       " 'insertion_size_limit',\n",
       " 'merge_identical',\n",
       " 'min_seq_len',\n",
       " 'optimize_flanking_regions',\n",
       " 'remove_seqs_with_wt_kmers',\n",
       " 'use_IDs',\n",
       " 'required_insertion_overlap_length',\n",
       " 'save_files',\n",
       " 'vcrs_strandedness'}"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allowable_kwargs_vk_build"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'dry_run',\n",
       " 'filtering_report_text_out',\n",
       " 'gtf',\n",
       " 'gtf_transcript_id_column',\n",
       " 'id_to_header_csv_out',\n",
       " 'k',\n",
       " 'max_ambiguous',\n",
       " 'mcrs_fasta_out',\n",
       " 'mcrs_t2g_out',\n",
       " 'mut_column',\n",
       " 'mut_id_column',\n",
       " 'mutations',\n",
       " 'mutations_updated_csv_out',\n",
       " 'out',\n",
       " 'overwrite',\n",
       " 'reference_out_dir',\n",
       " 'removed_variants_text_out',\n",
       " 'return_mutation_output',\n",
       " 'save_filtering_report_text',\n",
       " 'save_mutations_updated_csv',\n",
       " 'save_removed_variants_text',\n",
       " 'save_wt_mcrs_fasta_and_t2g',\n",
       " 'seq_id_column',\n",
       " 'sequences',\n",
       " 'store_full_sequences',\n",
       " 'translate',\n",
       " 'translate_end',\n",
       " 'translate_start',\n",
       " 'verbose',\n",
       " 'w',\n",
       " 'wt_mcrs_fasta_out',\n",
       " 'wt_mcrs_t2g_out'}"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "explicit_parameters_vk_build"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "varseek",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
