{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import varseek as vk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Required\n",
    "mutations = \"/home/jrich/data/varseek_data_fresh/reference/cosmic/CancerMutationCensus_AllData_Tsv_v100_GRCh37/CancerMutationCensus_AllData_v100_GRCh37_mutation_workflow_with_cdna.csv\"  # \"cosmic_cmc\"  # file path to mutations csv/tsv file OR one of the supported databases - eg cosmic_cmc\n",
    "sequences = \"/home/jrich/data/varseek_data_fresh/reference/ensembl_grch37_release93/Homo_sapiens.GRCh37.cdna.all.fa\"  # cdna  # file path to reference sequence fasta file OR, only if 'mutations' is in the supported databases, one of the following options is also supported: \"cds\", \"cdna\", \"genome\", \"cdna_and_genome\" - sequences for gget mutate\n",
    "rnaseq_fastq_files = \"\"  # pass as a single string or as a list of strings (if paired-end, then pass each pair sequentially with R1 preceding R2)\n",
    "\n",
    "\n",
    "# Optional\n",
    "verbose=True\n",
    "w=54  # window size for varseek build (should be at least 1 less than kallisto k)\n",
    "k=59\n",
    "threads = 32\n",
    "remove_Ns = True\n",
    "strand = \"unstranded\"  # \"forward\", \"reverse\", or \"unstranded\"\n",
    "strandedness = False  # strandedness for gget mutate and the building of the kb index (True = strandedness matters i.e., treat f and rc as 2 different sequences; False = strandedness does not matter i.e., treat f and rc as the same \n",
    "\n",
    "# vk build\n",
    "cosmic_version = 100  # COSMIC version for gget cosmic\n",
    "# os.environ['COSMIC_EMAIL'] = 'your_email'  # to avoid being prompted for email in varseek build\n",
    "# os.environ['COSMIC_PASSWORD'] = 'your_password'  # to avoid being prompted for password in varseek build\n",
    "\n",
    "# vk info\n",
    "columns_to_include=\"all\"\n",
    "dlist_reference_source = \"ensembl_grch37_release93\"  # ensembl_grchNUMBER_releaseNUMBER or t2t - eg ensembl_grch37_release93\n",
    "near_splice_junction_threshold=10\n",
    "save_exploded_df=True\n",
    "\n",
    "# vk filter\n",
    "fasta_filters = [\n",
    "    \"dlist_substring-equal=none\",  # filter out mutations which are a substring of the reference genome\n",
    "    \"pseudoaligned_to_human_reference_despite_not_truly_aligning-isnottrue\",  # filter out mutations which pseudoaligned to human genome despite not truly aligning\n",
    "    \"dlist-equal=none\",  #*** erase eventually when I want to d-list  # filter out mutations which are capable of being d-listed (given that I filter out the substrings above)\n",
    "    \"number_of_kmers_with_overlap_to_other_mcrs_items_in_mcrs_reference-max=999999\",  # filter out mutations which overlap with other MCRSs in the reference\n",
    "    \"number_of_mcrs_items_with_overlapping_kmers_in_mcrs_reference-max=999999\",  # filter out mutations which overlap with other MCRSs in the reference\n",
    "    \"longest_homopolymer_length-max=999999\",  # filters out MCRSs with repeating single nucleotide - eg 6\n",
    "    \"triplet_complexity-min=0\"  # filters out MCRSs with repeating triplets - eg 0.2\n",
    "]\n",
    "\n",
    "# kb ref\n",
    "dlist_kb_argument = \"None\"  # path to dlist fasta file or \"None\" (including the quotes)\n",
    "\n",
    "# vk fastqpp\n",
    "trim_edges_off_reads = False\n",
    "minimum_base_quality_trim_reads=13\n",
    "qualified_quality_phred = None\n",
    "unqualified_percent_limit = None\n",
    "n_base_limit = None\n",
    "length_required = None\n",
    "replace_low_quality_bases_with_N = False\n",
    "minimum_base_quality_replace_with_N=13\n",
    "split_reads_by_Ns = False\n",
    "run_fastqc = False\n",
    "\n",
    "# kb count\n",
    "assay = \"bulk\"  # \"bulk\" or \"sc\"  #!!! implement\n",
    "parity = \"single\"  # single or paired\n",
    "\n",
    "# vk clean\n",
    "minimum_count_filter = None  #!!! still need to tune this  \n",
    "use_binary_matrix = False\n",
    "drop_zero_columns = False\n",
    "filter_cells_by_min_counts = True  # True for auto-filter, numeric for set filter, None/False otherwise\n",
    "filter_cells_by_min_genes = 200\n",
    "filter_genes_by_min_cells = 3\n",
    "filter_cells_by_max_mt_content = False\n",
    "doublet_detection = True\n",
    "remove_doublets = False\n",
    "do_cpm_normalization = True\n",
    "split_reads_by_Ns = False\n",
    "dlist_file = None\n",
    "mutation_metadata_df_columns = None  #!!! replace\n",
    "\n",
    "\n",
    "# Paths\n",
    "out_dir_base = \"/home/jrich/data/varseek_data_fresh\"\n",
    "run_name = \"vk_build_pipeline_grch37_nov12\"\n",
    "\n",
    "# vk info\n",
    "bowtie_path=\"/home/jrich/opt/bowtie2-2.5.4/bowtie2-2.5.4-linux-x86_64\"\n",
    "mutations_csv=\"/home/jrich/data/varseek_data_fresh/reference/cosmic/CancerMutationCensus_AllData_Tsv_v100_GRCh37/CancerMutationCensus_AllData_v100_GRCh37_with_cdna.csv\"\n",
    "reference_cdna_fasta=\"/home/jrich/data/varseek_data_fresh/reference/ensembl_grch37_release93/Homo_sapiens.GRCh37.cdna.all.fa\"  # the one that matches up to mutation df annotations\n",
    "reference_genome_fasta=\"/home/jrich/data/varseek_data_fresh/reference/ensembl_grch37_release93/Homo_sapiens.GRCh37.dna.primary_assembly.fa\"  # the one that matches up to mutation df annotations\n",
    "gtf_path=\"/home/jrich/data/varseek_data_fresh/reference/ensembl_grch37_release93/Homo_sapiens.GRCh37.87.gtf\"  # the one that matches up to mutation df annotations\n",
    "seqtk=\"seqtk\"\n",
    "fastp=\"fastp\"\n",
    "\n",
    "# vk clean\n",
    "mutation_metadata_df_path = None  #!!! replace\n",
    "standard_index = \"\"\n",
    "standard_t2g = \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vk_ref_output_dict = vk.ref()\n",
    "\n",
    "index = vk_ref_output_dict[\"index\"]\n",
    "t2g = vk_ref_output_dict[\"t2g\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vk_count_output_dict = vk.clean()\n",
    "print(vk_count_output_dict)"
   ]
  }
 ],
 "metadata": {
  "language_info": {
   "name": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
